// @clr/angular/esm2015/timeline/timeline.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, Input, HostBinding } from '@angular/core';
import { ClrTimelineLayout } from './enums/timeline-layout.enum';
import { TimelineIconAttributeService } from './providers/timeline-icon-attribute.service';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
let ClrTimeline = class ClrTimeline {
    constructor() {
        this.layout = ClrTimelineLayout.HORIZONTAL;
    }
    get isVertical() {
        return this.layout === ClrTimelineLayout.VERTICAL;
    }
};
ClrTimeline.ɵfac = function ClrTimeline_Factory(t) { return new (t || ClrTimeline)(); };
ClrTimeline.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTimeline, selectors: [["clr-timeline"]], hostVars: 4, hostBindings: function ClrTimeline_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-timeline", true)("clr-timeline-vertical", ctx.isVertical);
    } }, inputs: { layout: ["clrLayout", "layout"] }, features: [ɵngcc0.ɵɵProvidersFeature([TimelineIconAttributeService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTimeline_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
__decorate([
    Input('clrLayout')
], ClrTimeline.prototype, "layout", void 0);
__decorate([
    HostBinding('class.clr-timeline-vertical')
], ClrTimeline.prototype, "isVertical", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimeline, [{
        type: Component,
        args: [{
                selector: 'clr-timeline',
                template: `
    <ng-content></ng-content>
  `,
                host: { '[class.clr-timeline]': 'true' },
                providers: [TimelineIconAttributeService]
            }]
    }], function () { return []; }, { layout: [{
            type: Input,
            args: ['clrLayout']
        }], isVertical: [{
            type: HostBinding,
            args: ['class.clr-timeline-vertical']
        }] }); })();
export { ClrTimeline };
