// @clr/angular/esm2015/timeline/timeline-step.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Component, ContentChild, ElementRef, Inject, Input, PLATFORM_ID } from '@angular/core';
import { ClrTimelineStepState } from './enums/timeline-step-state.enum';
import { TimelineIconAttributeService } from './providers/timeline-icon-attribute.service';
import { ClrTimelineStepTitle } from './timeline-step-title';
import { isPlatformBrowser } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../icon/icon';
import * as ɵngcc3 from '../progress/spinner/spinner';

function ClrTimelineStep_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "clr-icon");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("shape", ctx_r0.iconShape)("aria-label", ctx_r0.iconAriaLabel)("aria-current", ctx_r0.iconAriaCurrent);
} }
function ClrTimelineStep_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "clr-spinner", 4);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("aria-label", ctx_r2.iconAriaLabel);
} }
const _c0 = [[["clr-timeline-step-header"]], [["clr-timeline-step-title"]], [["clr-timeline-step-description"]]];
const _c1 = ["clr-timeline-step-header", "clr-timeline-step-title", "clr-timeline-step-description"];
let ClrTimelineStep = class ClrTimelineStep {
    constructor(iconAttributeService, platformId) {
        this.iconAttributeService = iconAttributeService;
        this.platformId = platformId;
        this.state = ClrTimelineStepState.NOT_STARTED;
    }
    ngAfterContentInit() {
        if (this.stepTitle && isPlatformBrowser(this.platformId)) {
            this.stepTitleText = this.stepTitle.nativeElement.innerText;
        }
    }
    get iconAriaCurrent() {
        return this.state === ClrTimelineStepState.CURRENT;
    }
    get iconAriaLabel() {
        return this.iconAttributeService.getAriaLabel(this.state);
    }
    get iconShape() {
        return this.iconAttributeService.getIconShape(this.state);
    }
    get isProcessing() {
        return this.state === ClrTimelineStepState.PROCESSING;
    }
};
ClrTimelineStep.ɵfac = function ClrTimelineStep_Factory(t) { return new (t || ClrTimelineStep)(ɵngcc0.ɵɵdirectiveInject(TimelineIconAttributeService), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
ClrTimelineStep.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTimelineStep, selectors: [["clr-timeline-step"]], contentQueries: function ClrTimelineStep_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrTimelineStepTitle, true, ElementRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepTitle = _t.first);
    } }, hostVars: 2, hostBindings: function ClrTimelineStep_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-timeline-step", true);
    } }, inputs: { state: ["clrState", "state"] }, ngContentSelectors: _c1, decls: 9, vars: 3, consts: [[1, "clr-sr-only"], [4, "ngIf", "ngIfElse"], [1, "clr-timeline-step-body"], ["processing", ""], ["clrMedium", ""]], template: function ClrTimelineStep_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ClrTimelineStep_ng_container_3_Template, 2, 3, "ng-container", 1);
        ɵngcc0.ɵɵelementStart(4, "div", 2);
        ɵngcc0.ɵɵprojection(5, 1);
        ɵngcc0.ɵɵprojection(6, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, ClrTimelineStep_ng_template_7_Template, 1, 1, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.stepTitleText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isProcessing)("ngIfElse", _r1);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ClrIconCustomTag, ɵngcc3.ClrSpinner], encapsulation: 2 });
ClrTimelineStep.ctorParameters = () => [
    { type: TimelineIconAttributeService },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
__decorate([
    Input('clrState')
], ClrTimelineStep.prototype, "state", void 0);
__decorate([
    ContentChild(ClrTimelineStepTitle, { read: ElementRef })
], ClrTimelineStep.prototype, "stepTitle", void 0);
ClrTimelineStep = __decorate([ __param(1, Inject(PLATFORM_ID))
], ClrTimelineStep);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimelineStep, [{
        type: Component,
        args: [{
                selector: 'clr-timeline-step',
                template: `
        <ng-content select="clr-timeline-step-header"></ng-content>
        <span class="clr-sr-only">{{stepTitleText}}</span>
        <ng-container *ngIf="!isProcessing; else processing">
            <clr-icon
                    [attr.shape]="iconShape"
                    [attr.aria-label]="iconAriaLabel"
                    [attr.aria-current]="iconAriaCurrent">
            </clr-icon>
        </ng-container>
        <div class="clr-timeline-step-body">
            <ng-content select="clr-timeline-step-title"></ng-content>
            <ng-content select="clr-timeline-step-description"></ng-content>
        </div>

        <ng-template #processing>
            <clr-spinner clrMedium [attr.aria-label]="iconAriaLabel"></clr-spinner>
        </ng-template>
    `,
                host: { '[class.clr-timeline-step]': 'true' }
            }]
    }], function () { return [{ type: TimelineIconAttributeService }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { state: [{
            type: Input,
            args: ['clrState']
        }], stepTitle: [{
            type: ContentChild,
            args: [ClrTimelineStepTitle, { read: ElementRef }]
        }] }); })();
export { ClrTimelineStep };
