// @clr/angular/esm2015/wizard/providers/header-actions.service.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { WizardNavigationService } from './wizard-navigation.service';
import * as ɵngcc0 from '@angular/core';
let HeaderActionService = class HeaderActionService {
    // this service communicates information about the presence/display of header actions
    // across the wizard
    constructor(navService) {
        this.navService = navService;
    }
    get wizardHasHeaderActions() {
        const wizardHdrActions = this.wizardHeaderActions;
        if (!wizardHdrActions) {
            return false;
        }
        return wizardHdrActions.toArray().length > 0;
    }
    get currentPageHasHeaderActions() {
        return this.navService.currentPage ? this.navService.currentPage.hasHeaderActions : false;
    }
    get showWizardHeaderActions() {
        return !this.currentPageHasHeaderActions && this.wizardHasHeaderActions;
    }
    get displayHeaderActionsWrapper() {
        return this.currentPageHasHeaderActions || this.wizardHasHeaderActions;
    }
};
HeaderActionService.ɵfac = function HeaderActionService_Factory(t) { return new (t || HeaderActionService)(ɵngcc0.ɵɵinject(WizardNavigationService)); };
HeaderActionService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: HeaderActionService, factory: HeaderActionService.ɵfac });
HeaderActionService.ctorParameters = () => [
    { type: WizardNavigationService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HeaderActionService, [{
        type: Injectable
    }], function () { return [{ type: WizardNavigationService }]; }, null); })();
export { HeaderActionService };
