// @clr/angular/esm5/popover/signpost/signpost.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrFocusOnViewInitModule } from '../../utils/focus/focus-on-view-init/focus-on-view-init.module';
import { ClrSignpost } from './signpost';
import { ClrSignpostContent } from './signpost-content';
import { ClrSignpostTrigger } from './signpost-trigger';
import * as ɵngcc0 from '@angular/core';
export var CLR_SIGNPOST_DIRECTIVES = [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger];
var ClrSignpostModule = /** @class */ (function () {
    function ClrSignpostModule() {
    }
ClrSignpostModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrSignpostModule });
ClrSignpostModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrSignpostModule_Factory(t) { return new (t || ClrSignpostModule)(); }, imports: [[CommonModule, ClrIconModule, ClrFocusOnViewInitModule],
        ClrConditionalModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrSignpostModule, { declarations: function () { return [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger]; }, imports: function () { return [CommonModule, ClrIconModule, ClrFocusOnViewInitModule]; }, exports: function () { return [ClrSignpost, ClrSignpostContent, ClrSignpostTrigger,
        ClrConditionalModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrSignpostModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrFocusOnViewInitModule],
                declarations: [CLR_SIGNPOST_DIRECTIVES],
                exports: [CLR_SIGNPOST_DIRECTIVES, ClrConditionalModule]
            }]
    }], function () { return []; }, null); })();
    return ClrSignpostModule;
}());
export { ClrSignpostModule };
