// @clr/angular/esm2015/utils/drag-and-drop/providers/global-drag-mode.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Renderer2 } from '@angular/core';
// This service class adds and removes the "in-drag" class to the document body element
// through its public enter() and exit() methods.
import * as ɵngcc0 from '@angular/core';
let GlobalDragModeService = class GlobalDragModeService {
    constructor(renderer) {
        this.renderer = renderer;
    }
    enter() {
        this.renderer.addClass(document.body, 'in-drag');
    }
    exit() {
        this.renderer.removeClass(document.body, 'in-drag');
    }
};
GlobalDragModeService.ɵfac = function GlobalDragModeService_Factory(t) { return new (t || GlobalDragModeService)(ɵngcc0.ɵɵinject(ɵngcc0.Renderer2)); };
GlobalDragModeService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GlobalDragModeService, factory: GlobalDragModeService.ɵfac });
GlobalDragModeService.ctorParameters = () => [
    { type: Renderer2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GlobalDragModeService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, null); })();
export { GlobalDragModeService };
