// @clr/angular/esm2015/layout/nav/chocolate/nav-detection-oompa-loompa.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ChangeDetectorRef, Directive, Optional } from '@angular/core';
import { OompaLoompa } from '../../../utils/chocolate/oompa-loompa';
import { ResponsiveNavigationService } from '../providers/responsive-navigation.service';
import { MainContainerWillyWonka } from './main-container-willy-wonka';
import * as ɵngcc0 from '@angular/core';
let NavDetectionOompaLoompa = class NavDetectionOompaLoompa extends OompaLoompa {
    constructor(cdr, willyWonka, responsiveNavService) {
        if (!willyWonka) {
            throw new Error('clr-header should only be used inside of a clr-main-container');
        }
        super(cdr, willyWonka);
        this.responsiveNavService = responsiveNavService;
    }
    // NavDetectionOompaLoompa is the addition of the nav levels
    // Since we support 2 levels, the possibilities are 0, 1 or 3 (1 + 2)
    get flavor() {
        return this.responsiveNavService.responsiveNavList.reduce((sum, navLevel) => sum + navLevel, 0);
    }
};
NavDetectionOompaLoompa.ɵfac = function NavDetectionOompaLoompa_Factory(t) { return new (t || NavDetectionOompaLoompa)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(MainContainerWillyWonka, 8), ɵngcc0.ɵɵdirectiveInject(ResponsiveNavigationService)); };
NavDetectionOompaLoompa.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NavDetectionOompaLoompa, selectors: [["clr-header"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
NavDetectionOompaLoompa.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: MainContainerWillyWonka, decorators: [{ type: Optional }] },
    { type: ResponsiveNavigationService }
];
NavDetectionOompaLoompa = __decorate([ __param(1, Optional())
], NavDetectionOompaLoompa);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NavDetectionOompaLoompa, [{
        type: Directive,
        args: [{ selector: 'clr-header' }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: MainContainerWillyWonka, decorators: [{
                type: Optional
            }] }, { type: ResponsiveNavigationService }]; }, null); })();
export { NavDetectionOompaLoompa };
