// @clr/angular/esm5/button/button-loading/loading-button.module.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrLoadingButton } from './loading-button';
import * as ɵngcc0 from '@angular/core';
export var CLR_LOADING_BUTTON_DIRECTIVES = [ClrLoadingButton];
var ClrLoadingButtonModule = /** @class */ (function () {
    function ClrLoadingButtonModule() {
    }
ClrLoadingButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrLoadingButtonModule });
ClrLoadingButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrLoadingButtonModule_Factory(t) { return new (t || ClrLoadingButtonModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrLoadingButtonModule, { declarations: function () { return [ClrLoadingButton]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrLoadingButton]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrLoadingButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_LOADING_BUTTON_DIRECTIVES],
                exports: [CLR_LOADING_BUTTON_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrLoadingButtonModule;
}());
export { ClrLoadingButtonModule };
