// @clr/angular/esm2015/forms/select/select.module.js
/**
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrSelect } from './select';
import { ClrSelectContainer } from './select-container';
import * as ɵngcc0 from '@angular/core';
let ClrSelectModule = class ClrSelectModule {
};
ClrSelectModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrSelectModule });
ClrSelectModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrSelectModule_Factory(t) { return new (t || ClrSelectModule)(); }, imports: [[CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrSelectModule, { declarations: function () { return [ClrSelect, ClrSelectContainer]; }, imports: function () { return [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule]; }, exports: function () { return [ClrCommonFormsModule, ClrSelect, ClrSelectContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrSelectModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
                declarations: [ClrSelect, ClrSelectContainer],
                exports: [ClrCommonFormsModule, ClrSelect, ClrSelectContainer],
                entryComponents: [ClrSelectContainer]
            }]
    }], null, null); })();
export { ClrSelectModule };
