// @clr/angular/esm5/progress/progress-bars/progress-bar.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrProgressBar } from './progress-bar';
import * as ɵngcc0 from '@angular/core';
export var CLR_PROGRESS_BAR_DIRECTIVES = [ClrProgressBar];
var ClrProgressBarModule = /** @class */ (function () {
    function ClrProgressBarModule() {
    }
ClrProgressBarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrProgressBarModule });
ClrProgressBarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrProgressBarModule_Factory(t) { return new (t || ClrProgressBarModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrProgressBarModule, { declarations: function () { return [ClrProgressBar]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrProgressBar]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrProgressBarModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_PROGRESS_BAR_DIRECTIVES],
                exports: [CLR_PROGRESS_BAR_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrProgressBarModule;
}());
export { ClrProgressBarModule };
