// @clr/angular/esm2015/button/button-group/button-group.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Component, Inject, ContentChildren, Input } from '@angular/core';
import { ButtonInGroupService } from '../providers/button-in-group.service';
import { ClrCommonStringsService } from '../../utils/i18n/common-strings.service';
import { UNIQUE_ID, UNIQUE_ID_PROVIDER } from '../../utils/id-generator/id-generator.service';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import { ClrPopoverEventsService } from '../../utils/popover/providers/popover-events.service';
import { ClrPopoverPositionService } from '../../utils/popover/providers/popover-position.service';
import { ClrPopoverPositions } from '../../utils/popover/enums/positions.enum';
import { ClrButton } from './button';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../utils/popover/popover-anchor';
import * as ɵngcc3 from '../../utils/popover/popover-open-close-button';
import * as ɵngcc4 from '../../icon/icon';
import * as ɵngcc5 from '../../utils/popover/popover-content';

function ClrButtonGroup_ng_container_0_ng_template_1_Template(rf, ctx) { }
function ClrButtonGroup_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ClrButtonGroup_ng_container_0_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const inlineButton_r4 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", inlineButton_r4.templateRef);
} }
function ClrButtonGroup_ng_container_1_div_5_ng_template_1_Template(rf, ctx) { }
function ClrButtonGroup_ng_container_1_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵtemplate(1, ClrButtonGroup_ng_container_1_div_5_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵproperty("id", ctx_r7.popoverId);
    ɵngcc0.ɵɵattribute("id", ctx_r7.popoverId)("aria-hidden", !ctx_r7.open);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function ClrButtonGroup_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 4, 5);
    ɵngcc0.ɵɵelementStart(3, "button", 6);
    ɵngcc0.ɵɵelement(4, "clr-icon", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, ClrButtonGroup_ng_container_1_div_5_Template, 2, 4, "div", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.menuPosition);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("aria-controls", ctx_r1.popoverId)("aria-expanded", ctx_r1.open)("aria-label", ctx_r1.commonStrings.keys.rowActions);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("title", ctx_r1.commonStrings.keys.more);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("clrPopoverContent", ctx_r1.open)("clrPopoverContentAt", ctx_r1.popoverPosition)("clrPopoverContentOutsideClickToClose", true)("clrPopoverContentScrollToClose", true);
} }
function ClrButtonGroup_ng_template_2_ng_container_0_ng_template_1_Template(rf, ctx) { }
function ClrButtonGroup_ng_template_2_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ClrButtonGroup_ng_template_2_ng_container_0_ng_template_1_Template, 0, 0, "ng-template", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const menuButton_r10 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", menuButton_r10.templateRef);
} }
function ClrButtonGroup_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ClrButtonGroup_ng_template_2_ng_container_0_Template, 2, 1, "ng-container", 0);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.menuButtons);
} }
let ClrButtonGroup = class ClrButtonGroup {
    constructor(buttonGroupNewService, toggleService, popoverId, commonStrings) {
        this.buttonGroupNewService = buttonGroupNewService;
        this.toggleService = toggleService;
        this.popoverId = popoverId;
        this.commonStrings = commonStrings;
        this.popoverPosition = ClrPopoverPositions['bottom-left'];
        this.inlineButtons = [];
        this.menuButtons = [];
    }
    get open() {
        return this.toggleService.open;
    }
    /**
     * 1. Initializes the initial Button Group View
     * 2. Subscribes to changes on the ContentChildren
     *    in case the user content projection changes
     */
    ngAfterContentInit() {
        this.initializeButtons();
        this.buttonGroupNewService.changes.subscribe(button => this.rearrangeButton(button));
        this.buttons.changes.subscribe(() => {
            this.initializeButtons();
        });
    }
    /**
     * Moves the button into the other ViewContainer
     * when an update is received.
     *
     * @param button
     */
    rearrangeButton(button) {
        let fromView;
        let toView;
        if (button.inMenu) {
            fromView = this.inlineButtons;
            toView = this.menuButtons;
        }
        else {
            fromView = this.menuButtons;
            toView = this.inlineButtons;
        }
        const index = fromView.indexOf(button);
        if (index > -1) {
            fromView.splice(index, 1);
            const moveIndex = this.getMoveIndex(button);
            if (moveIndex <= toView.length) {
                toView.splice(moveIndex, 0, button);
            }
        }
    }
    /**
     * Author: Eudes
     *
     * Finds the order of a button w.r.t other buttons
     *
     * @param buttonToMove
     * @returns
     */
    getMoveIndex(buttonToMove) {
        const tempArr = this.buttons.filter(button => button.inMenu === buttonToMove.inMenu);
        return tempArr.indexOf(buttonToMove);
    }
    initializeButtons() {
        const tempInlineButtons = [];
        const tempInMenuButtons = [];
        this.buttons.forEach(button => {
            if (button.inMenu) {
                tempInMenuButtons.push(button);
            }
            else {
                tempInlineButtons.push(button);
            }
        });
        this.inlineButtons = tempInlineButtons;
        this.menuButtons = tempInMenuButtons;
    }
    get menuPosition() {
        return this._menuPosition;
    }
    set menuPosition(pos) {
        if (pos && ClrPopoverPositions[pos]) {
            this._menuPosition = pos;
        }
        else {
            this._menuPosition = 'bottom-left';
        }
        this.popoverPosition = ClrPopoverPositions[this._menuPosition];
    }
};
ClrButtonGroup.ɵfac = function ClrButtonGroup_Factory(t) { return new (t || ClrButtonGroup)(ɵngcc0.ɵɵdirectiveInject(ButtonInGroupService), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(UNIQUE_ID), ɵngcc0.ɵɵdirectiveInject(ClrCommonStringsService)); };
ClrButtonGroup.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrButtonGroup, selectors: [["clr-button-group"]], contentQueries: function ClrButtonGroup_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrButton, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.buttons = _t);
    } }, hostVars: 2, hostBindings: function ClrButtonGroup_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("btn-group", true);
    } }, inputs: { menuPosition: ["clrMenuPosition", "menuPosition"] }, features: [ɵngcc0.ɵɵProvidersFeature([
            ButtonInGroupService,
            UNIQUE_ID_PROVIDER,
            ClrPopoverToggleService,
            ClrPopoverEventsService,
            ClrPopoverPositionService,
        ])], decls: 4, vars: 2, consts: [[4, "ngFor", "ngForOf"], [4, "ngIf"], ["ref", ""], [3, "ngTemplateOutlet"], [1, "btn-group-overflow", "open", 3, "ngClass"], ["anchor", ""], ["clrPopoverAnchor", "", "clrPopoverOpenCloseButton", "", 1, "btn", "dropdown-toggle"], ["shape", "ellipsis-horizontal"], ["role", "menu", "class", "dropdown-menu clr-button-group-menu", 3, "id", 4, "clrPopoverContent", "clrPopoverContentAt", "clrPopoverContentOutsideClickToClose", "clrPopoverContentScrollToClose"], ["role", "menu", 1, "dropdown-menu", "clr-button-group-menu", 3, "id"]], template: function ClrButtonGroup_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ClrButtonGroup_ng_container_0_Template, 2, 1, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, ClrButtonGroup_ng_container_1_Template, 6, 9, "ng-container", 1);
        ɵngcc0.ɵɵtemplate(2, ClrButtonGroup_ng_template_2_Template, 1, 1, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.inlineButtons);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.menuButtons.length > 0);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgClass, ɵngcc2.ClrPopoverAnchor, ɵngcc3.ClrPopoverOpenCloseButton, ɵngcc4.ClrIconCustomTag, ɵngcc5.ClrPopoverContent], encapsulation: 2 });
ClrButtonGroup.ctorParameters = () => [
    { type: ButtonInGroupService },
    { type: ClrPopoverToggleService },
    { type: String, decorators: [{ type: Inject, args: [UNIQUE_ID,] }] },
    { type: ClrCommonStringsService }
];
__decorate([
    ContentChildren(ClrButton)
], ClrButtonGroup.prototype, "buttons", void 0);
__decorate([
    Input('clrMenuPosition')
], ClrButtonGroup.prototype, "menuPosition", null);
ClrButtonGroup = __decorate([ __param(2, Inject(UNIQUE_ID))
], ClrButtonGroup);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrButtonGroup, [{
        type: Component,
        args: [{
                selector: 'clr-button-group',
                template: "<!--\n  ~ Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.\n  ~ This software is released under MIT license.\n  ~ The full license information can be found in LICENSE in the root directory of this project.\n  -->\n\n<ng-container *ngFor=\"let inlineButton of inlineButtons\">\n    <ng-template [ngTemplateOutlet]=\"inlineButton.templateRef\"></ng-template>\n</ng-container>\n<ng-container *ngIf=\"menuButtons.length > 0\">\n    <div\n        class=\"btn-group-overflow open\"\n        [ngClass]=\"menuPosition\"\n        #anchor>\n        <button\n            class=\"btn dropdown-toggle\"\n            clrPopoverAnchor\n            clrPopoverOpenCloseButton\n            [attr.aria-controls]=\"popoverId\"\n            [attr.aria-expanded]=\"open\"\n            [attr.aria-label]=\"commonStrings.keys.rowActions\">\n            <clr-icon shape=\"ellipsis-horizontal\" [attr.title]=\"commonStrings.keys.more\"></clr-icon>\n        </button>\n        <div\n            role=\"menu\" \n            class=\"dropdown-menu clr-button-group-menu\"\n            [id]=\"popoverId\"\n            [attr.id]=\"popoverId\"\n            [attr.aria-hidden]=\"!open\"\n            *clrPopoverContent=\"open at popoverPosition; outsideClickToClose: true; scrollToClose: true\">\n            <ng-template [ngTemplateOutlet]=\"ref\"></ng-template>\n        </div>\n    </div>\n</ng-container>\n<ng-template #ref>\n    <ng-container *ngFor=\"let menuButton of menuButtons\">\n        <ng-template [ngTemplateOutlet]=\"menuButton.templateRef\"></ng-template>\n    </ng-container>\n</ng-template>\n",
                providers: [
                    ButtonInGroupService,
                    UNIQUE_ID_PROVIDER,
                    ClrPopoverToggleService,
                    ClrPopoverEventsService,
                    ClrPopoverPositionService,
                ],
                host: { '[class.btn-group]': 'true' }
            }]
    }], function () { return [{ type: ButtonInGroupService }, { type: ClrPopoverToggleService }, { type: String, decorators: [{
                type: Inject,
                args: [UNIQUE_ID]
            }] }, { type: ClrCommonStringsService }]; }, { menuPosition: [{
            type: Input,
            args: ['clrMenuPosition']
        }], buttons: [{
            type: ContentChildren,
            args: [ClrButton]
        }] }); })();
export { ClrButtonGroup };
