// @clr/angular/esm2015/utils/conditional/if-expanded.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { ClrLoadingState } from '../loading/loading';
import * as ɵngcc0 from '@angular/core';
let IfExpandService = class IfExpandService {
    constructor() {
        this.expandable = 0;
        this._loading = false;
        this._expanded = false;
        this._expandChange = new Subject();
    }
    get loading() {
        return this._loading;
    }
    set loading(value) {
        value = !!value;
        if (value !== this._loading) {
            this._loading = value;
        }
    }
    get expanded() {
        return this._expanded;
    }
    set expanded(value) {
        value = !!value;
        if (value !== this._expanded) {
            this._expanded = value;
            this._expandChange.next(value);
        }
    }
    toggle() {
        this.expanded = !this._expanded;
    }
    get expandChange() {
        return this._expandChange.asObservable();
    }
    loadingStateChange(state) {
        switch (state) {
            case ClrLoadingState.LOADING:
                this.loading = true;
                break;
            default:
                this.loading = false;
                break;
        }
    }
};
IfExpandService.ɵfac = function IfExpandService_Factory(t) { return new (t || IfExpandService)(); };
IfExpandService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: IfExpandService, factory: IfExpandService.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IfExpandService, [{
        type: Injectable
    }], function () { return []; }, null); })();
export { IfExpandService };
