// @clr/angular/esm2015/utils/conditional/if-open.directive.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, EventEmitter, Input, OnDestroy, Output, TemplateRef, ViewContainerRef } from '@angular/core';
import { ClrPopoverToggleService } from '../popover/providers/popover-toggle.service';
import * as ɵngcc0 from '@angular/core';
let ClrIfOpen = 
/**********
 *
 * @class ClrIfOpen
 *
 * @description
 * A structural directive that controls whether or not the associated TemplateRef is instantiated or not.
 * It makes use of a Component instance level service: ClrPopoverToggleService to maintain state between itself and the component
 * using it in the component template.
 *
 */
class ClrIfOpen {
    constructor(toggleService, template, container) {
        this.toggleService = toggleService;
        this.template = template;
        this.container = container;
        /**********
         * @property openChange
         *
         * @description
         * An event emitter that emits when the open property is set to allow for 2way binding when the directive is
         * used with de-structured / de-sugared syntax.
         */
        this.openChange = new EventEmitter(false);
        this.subscription = this.toggleService.openChange.subscribe(change => {
            this.updateView(change);
            this.openChange.emit(change);
        });
    }
    /*********
     *
     * @description
     * A setter that updates ClrPopoverToggleService.open with value.
     *
     * @param value
     */
    set open(value) {
        this.toggleService.open = value;
    }
    /********
     *
     * @description
     * A getter that returns the current ClrPopoverToggleService.open value.
     *
     */
    get open() {
        return this.toggleService.open;
    }
    /*********
     *
     * @description
     * Function that takes a boolean value and either created an embedded view for the associated ViewContainerRef or,
     * Clears all views from the ViewContainerRef
     * @param value
     */
    updateView(value) {
        if (value) {
            this.container.createEmbeddedView(this.template);
        }
        else {
            this.container.clear();
        }
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
};
ClrIfOpen.ɵfac = function ClrIfOpen_Factory(t) { return new (t || ClrIfOpen)(ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ClrIfOpen.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrIfOpen, selectors: [["", "clrIfOpen", ""]], inputs: { open: ["clrIfOpen", "open"] }, outputs: { openChange: "clrIfOpenChange" } });
ClrIfOpen.ctorParameters = () => [
    { type: ClrPopoverToggleService },
    { type: TemplateRef },
    { type: ViewContainerRef }
];
__decorate([
    Input('clrIfOpen')
], ClrIfOpen.prototype, "open", null);
__decorate([
    Output('clrIfOpenChange')
], ClrIfOpen.prototype, "openChange", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrIfOpen, [{
        type: Directive,
        args: [{ selector: '[clrIfOpen]' }]
    }], function () { return [{ type: ClrPopoverToggleService }, { type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }]; }, { openChange: [{
            type: Output,
            args: ['clrIfOpenChange']
        }], open: [{
            type: Input,
            args: ['clrIfOpen']
        }] }); })();
export { ClrIfOpen };
