// @clr/angular/esm5/layout/vertical-nav/providers/vertical-nav-group-registration.service.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var VerticalNavGroupRegistrationService = /** @class */ (function () {
    function VerticalNavGroupRegistrationService() {
        this.navGroupCount = 0;
    }
    VerticalNavGroupRegistrationService.prototype.registerNavGroup = function () {
        this.navGroupCount++;
    };
    VerticalNavGroupRegistrationService.prototype.unregisterNavGroup = function () {
        this.navGroupCount--;
    };
VerticalNavGroupRegistrationService.ɵfac = function VerticalNavGroupRegistrationService_Factory(t) { return new (t || VerticalNavGroupRegistrationService)(); };
VerticalNavGroupRegistrationService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: VerticalNavGroupRegistrationService, factory: function (t) { return VerticalNavGroupRegistrationService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VerticalNavGroupRegistrationService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return VerticalNavGroupRegistrationService;
}());
export { VerticalNavGroupRegistrationService };
