// @clr/angular/esm5/forms/input/input.module.js
/**
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrInput } from './input';
import { ClrInputContainer } from './input-container';
import * as ɵngcc0 from '@angular/core';
var ClrInputModule = /** @class */ (function () {
    function ClrInputModule() {
    }
ClrInputModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrInputModule });
ClrInputModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrInputModule_Factory(t) { return new (t || ClrInputModule)(); }, imports: [[CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrInputModule, { declarations: function () { return [ClrInput, ClrInputContainer]; }, imports: function () { return [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule]; }, exports: function () { return [ClrCommonFormsModule, ClrInput, ClrInputContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrInputModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
                declarations: [ClrInput, ClrInputContainer],
                exports: [ClrCommonFormsModule, ClrInput, ClrInputContainer],
                entryComponents: [ClrInputContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrInputModule;
}());
export { ClrInputModule };
