// @clr/angular/esm2015/forms/common/if-error/if-error.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Input, Optional, TemplateRef, ViewContainerRef } from '@angular/core';
import { IfErrorService } from './if-error.service';
import { NgControlService } from '../providers/ng-control.service';
import * as ɵngcc0 from '@angular/core';
let ClrIfError = class ClrIfError {
    constructor(ifErrorService, ngControlService, template, container) {
        this.ifErrorService = ifErrorService;
        this.ngControlService = ngControlService;
        this.template = template;
        this.container = container;
        this.subscriptions = [];
        this.displayed = false;
        if (!this.ifErrorService) {
            throw new Error('clrIfError can only be used within a form control container element like clr-input-container');
        }
        else {
            this.displayError(false);
        }
        this.subscriptions.push(this.ngControlService.controlChanges.subscribe(control => {
            this.control = control;
        }));
        this.subscriptions.push(this.ifErrorService.statusChanges.subscribe(invalid => {
            // If there is a specific error to track, check it, otherwise check overall validity
            if (this.error && this.control) {
                this.displayError(this.control.hasError(this.error));
            }
            else {
                this.displayError(invalid);
            }
        }));
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
    displayError(invalid) {
        if (invalid && !this.displayed) {
            this.container.createEmbeddedView(this.template, { error: this.control.getError(this.error) });
            this.displayed = true;
        }
        else if (!invalid) {
            this.container.clear();
            this.displayed = false;
        }
    }
};
ClrIfError.ɵfac = function ClrIfError_Factory(t) { return new (t || ClrIfError)(ɵngcc0.ɵɵdirectiveInject(IfErrorService, 8), ɵngcc0.ɵɵdirectiveInject(NgControlService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ClrIfError.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrIfError, selectors: [["", "clrIfError", ""]], inputs: { error: ["clrIfError", "error"] } });
ClrIfError.ctorParameters = () => [
    { type: IfErrorService, decorators: [{ type: Optional }] },
    { type: NgControlService, decorators: [{ type: Optional }] },
    { type: TemplateRef },
    { type: ViewContainerRef }
];
__decorate([
    Input('clrIfError')
], ClrIfError.prototype, "error", void 0);
ClrIfError = __decorate([ __param(0, Optional()),
    __param(1, Optional())
], ClrIfError);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrIfError, [{
        type: Directive,
        args: [{ selector: '[clrIfError]' }]
    }], function () { return [{ type: IfErrorService, decorators: [{
                type: Optional
            }] }, { type: NgControlService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }]; }, { error: [{
            type: Input,
            args: ['clrIfError']
        }] }); })();
export { ClrIfError };
