// @clr/angular/esm2015/forms/common/if-error/if-error.service.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { NgControlService } from '../providers/ng-control.service';
import * as ɵngcc0 from '@angular/core';
let IfErrorService = class IfErrorService {
    constructor(ngControlService) {
        this.ngControlService = ngControlService;
        // Implement our own status changes observable, since Angular controls don't
        // fire on events like blur, and we want to return the boolean state instead of a string
        this._statusChanges = new Subject();
        this.subscriptions = [];
        // Wait for the control to be available
        this.subscriptions.push(this.ngControlService.controlChanges.subscribe(control => {
            if (control) {
                this.control = control;
                this.listenForChanges();
            }
        }));
    }
    get statusChanges() {
        return this._statusChanges.asObservable();
    }
    // Subscribe to the status change events, only after touched and emit the control
    listenForChanges() {
        this.subscriptions.push(this.control.statusChanges.subscribe(() => {
            this.sendValidity();
        }));
    }
    sendValidity() {
        if (this.control.touched && this.control.invalid) {
            this._statusChanges.next(true);
        }
        else {
            this._statusChanges.next(false);
        }
    }
    // Allows a control to push a status check upstream, such as on blur
    triggerStatusChange() {
        if (this.control) {
            this.sendValidity();
        }
    }
    // Clean up subscriptions
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
};
IfErrorService.ɵfac = function IfErrorService_Factory(t) { return new (t || IfErrorService)(ɵngcc0.ɵɵinject(NgControlService)); };
IfErrorService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: IfErrorService, factory: IfErrorService.ɵfac });
IfErrorService.ctorParameters = () => [
    { type: NgControlService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IfErrorService, [{
        type: Injectable
    }], function () { return [{ type: NgControlService }]; }, null); })();
export { IfErrorService };
