// @clr/angular/esm2015/forms/common/form.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { ContentChildren, Directive, ElementRef, HostListener, Inject, PLATFORM_ID, QueryList, Input, } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { LayoutService } from './providers/layout.service';
import { MarkControlService } from './providers/mark-control.service';
import { ClrCommonStringsService } from '../../utils/i18n/common-strings.service';
import { ClrLabel } from './label';
import { ClrAriaLiveService } from '../../utils/a11y/aria-live.service';
import * as ɵngcc0 from '@angular/core';
let ClrForm = class ClrForm {
    constructor(layoutService, markControlService, platformId, el, commonStrings, ariaLiveService) {
        this.layoutService = layoutService;
        this.markControlService = markControlService;
        this.platformId = platformId;
        this.el = el;
        this.commonStrings = commonStrings;
        this.ariaLiveService = ariaLiveService;
        this.invalidControls = [];
    }
    set labelSize(size) {
        this.layoutService.labelSize = size;
    }
    /** @deprecated since 2.0 */
    markAsDirty(updateAriaLiveText) {
        this.markAsTouched((updateAriaLiveText = true));
    }
    // Trying to avoid adding an input and keep this backwards compatible at the same time
    markAsTouched(updateAriaLiveText) {
        this.markControlService.markAsTouched();
        // I don't think consumers will call this with undefined, null or other values but
        // want to make sure this only guards against when this is called with false
        if (updateAriaLiveText !== false && isPlatformBrowser(this.platformId)) {
            this.invalidControls = Array.from(this.el.nativeElement.querySelectorAll('.ng-invalid'));
            if (this.invalidControls.length > 0) {
                this.invalidControls[0].focus();
                this.updateAriaLive();
            }
        }
    }
    onFormSubmit() {
        this.markAsTouched();
    }
    /** @deprecated since 3.0, remove in 4.0 */
    updateAriaLive() {
        if (this.invalidControls.length === 0) {
            return;
        }
        const errorList = this.labels.filter(label => this.invalidControls.find(control => label.forAttr === control.id));
        this.ariaLiveService.announce(this.commonStrings.parse(this.commonStrings.keys.formErrorSummary, { ERROR_NUMBER: errorList.length.toString() }));
    }
};
ClrForm.ɵfac = function ClrForm_Factory(t) { return new (t || ClrForm)(ɵngcc0.ɵɵdirectiveInject(LayoutService), ɵngcc0.ɵɵdirectiveInject(MarkControlService), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ClrCommonStringsService), ɵngcc0.ɵɵdirectiveInject(ClrAriaLiveService)); };
ClrForm.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrForm, selectors: [["", "clrForm", ""]], contentQueries: function ClrForm_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrLabel, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.labels = _t);
    } }, hostVars: 6, hostBindings: function ClrForm_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("submit", function ClrForm_submit_HostBindingHandler() { return ctx.onFormSubmit(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-form", true)("clr-form-horizontal", ctx.layoutService.isHorizontal())("clr-form-compact", ctx.layoutService.isCompact());
    } }, inputs: { labelSize: ["clrLabelSize", "labelSize"] }, features: [ɵngcc0.ɵɵProvidersFeature([LayoutService, MarkControlService, ClrAriaLiveService])] });
ClrForm.ctorParameters = () => [
    { type: LayoutService },
    { type: MarkControlService },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
    { type: ElementRef },
    { type: ClrCommonStringsService },
    { type: ClrAriaLiveService }
];
__decorate([
    Input('clrLabelSize')
], ClrForm.prototype, "labelSize", null);
__decorate([
    ContentChildren(ClrLabel, { descendants: true })
], ClrForm.prototype, "labels", void 0);
__decorate([
    HostListener('submit')
], ClrForm.prototype, "onFormSubmit", null);
ClrForm = __decorate([ __param(2, Inject(PLATFORM_ID))
], ClrForm);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrForm, [{
        type: Directive,
        args: [{
                selector: '[clrForm]',
                providers: [LayoutService, MarkControlService, ClrAriaLiveService],
                host: {
                    '[class.clr-form]': 'true',
                    '[class.clr-form-horizontal]': 'layoutService.isHorizontal()',
                    '[class.clr-form-compact]': 'layoutService.isCompact()'
                }
            }]
    }], function () { return [{ type: LayoutService }, { type: MarkControlService }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }, { type: ɵngcc0.ElementRef }, { type: ClrCommonStringsService }, { type: ClrAriaLiveService }]; }, { labelSize: [{
            type: Input,
            args: ['clrLabelSize']
        }], onFormSubmit: [{
            type: HostListener,
            args: ['submit']
        }], labels: [{
            type: ContentChildren,
            args: [ClrLabel, { descendants: true }]
        }] }); })();
export { ClrForm };
