// @clr/angular/esm2015/forms/common/layout.js
/**
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { Layouts, LayoutService } from './providers/layout.service';
import * as ɵngcc0 from '@angular/core';
let ClrLayout = class ClrLayout {
    constructor(layoutService) {
        this.layoutService = layoutService;
    }
    ngOnInit() {
        // Only set the layout if it is a valid option
        if (this.layout && this.layoutService.isValid(this.layout)) {
            this.layoutService.layout = this.layout;
        }
    }
};
ClrLayout.ɵfac = function ClrLayout_Factory(t) { return new (t || ClrLayout)(ɵngcc0.ɵɵdirectiveInject(LayoutService)); };
ClrLayout.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrLayout, selectors: [["", "clrForm", "", "clrLayout", ""]], inputs: { layout: ["clrLayout", "layout"] } });
ClrLayout.ctorParameters = () => [
    { type: LayoutService }
];
__decorate([
    Input('clrLayout')
], ClrLayout.prototype, "layout", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrLayout, [{
        type: Directive,
        args: [{
                selector: '[clrForm][clrLayout]'
            }]
    }], function () { return [{ type: LayoutService }]; }, { layout: [{
            type: Input,
            args: ['clrLayout']
        }] }); })();
export { ClrLayout };
