// @clr/angular/esm2015/layout/vertical-nav/vertical-nav-link.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, HostListener, Optional } from '@angular/core';
import { VerticalNavGroupService } from './providers/vertical-nav-group.service';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["clrVerticalNavLink", ""];
const _c1 = [[["", "clrVerticalNavIcon", ""]], "*"];
const _c2 = ["[clrVerticalNavIcon]", "*"];
let ClrVerticalNavLink = class ClrVerticalNavLink {
    constructor(_navGroupService) {
        this._navGroupService = _navGroupService;
    }
    expandParentNavGroup() {
        if (this._navGroupService) {
            this._navGroupService.expand();
        }
    }
};
ClrVerticalNavLink.ɵfac = function ClrVerticalNavLink_Factory(t) { return new (t || ClrVerticalNavLink)(ɵngcc0.ɵɵdirectiveInject(VerticalNavGroupService, 8)); };
ClrVerticalNavLink.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrVerticalNavLink, selectors: [["", "clrVerticalNavLink", ""]], hostAttrs: [1, "nav-link"], hostBindings: function ClrVerticalNavLink_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrVerticalNavLink_click_HostBindingHandler() { return ctx.expandParentNavGroup(); });
    } }, attrs: _c0, ngContentSelectors: _c2, decls: 3, vars: 0, consts: [[1, "nav-text"]], template: function ClrVerticalNavLink_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵprojection(2, 1);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
ClrVerticalNavLink.ctorParameters = () => [
    { type: VerticalNavGroupService, decorators: [{ type: Optional }] }
];
__decorate([
    HostListener('click')
], ClrVerticalNavLink.prototype, "expandParentNavGroup", null);
ClrVerticalNavLink = __decorate([ __param(0, Optional())
], ClrVerticalNavLink);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrVerticalNavLink, [{
        type: Component,
        args: [{
                selector: '[clrVerticalNavLink]',
                template: `
        <ng-content select="[clrVerticalNavIcon]"></ng-content>
        <span class="nav-text">
            <ng-content></ng-content>    
        </span>
    `,
                host: { class: 'nav-link' }
            }]
    }], function () { return [{ type: VerticalNavGroupService, decorators: [{
                type: Optional
            }] }]; }, { expandParentNavGroup: [{
            type: HostListener,
            args: ['click']
        }] }); })();
export { ClrVerticalNavLink };
