// @clr/angular/esm5/emphasis/alert/alerts-pager.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { MultiAlertService } from './providers/multi-alert.service';
import { ClrCommonStringsService } from '../../utils/i18n/common-strings.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../icon/icon';
var ClrAlertsPager = /** @class */ (function () {
    function ClrAlertsPager(multiAlertService, commonStrings) {
        this.multiAlertService = multiAlertService;
        this.commonStrings = commonStrings;
        this.currentAlertChange = new EventEmitter(false);
        this.currentAlertIndexChange = new EventEmitter();
    }
    Object.defineProperty(ClrAlertsPager.prototype, "currentAlert", {
        get: function () {
            return this.multiAlertService.currentAlert;
        },
        /**
         * Input/Output to support two way binding on current alert instance
         */
        set: function (alert) {
            if (alert) {
                this.multiAlertService.currentAlert = alert;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrAlertsPager.prototype, "currentAlertIndex", {
        get: function () {
            return this.multiAlertService.current;
        },
        /**
         * Input/Output to support two way binding on current alert index
         */
        set: function (index) {
            this.multiAlertService.current = index;
        },
        enumerable: true,
        configurable: true
    });
    ClrAlertsPager.prototype.ngOnInit = function () {
        var _this = this;
        this.multiAlertServiceChanges = this.multiAlertService.changes.subscribe(function (index) {
            _this.currentAlertIndexChange.emit(index);
            _this.currentAlertChange.emit(_this.multiAlertService.activeAlerts[index]);
        });
    };
    ClrAlertsPager.prototype.pageUp = function () {
        this.multiAlertService.next();
    };
    ClrAlertsPager.prototype.pageDown = function () {
        this.multiAlertService.previous();
    };
    ClrAlertsPager.prototype.ngOnDestroy = function () {
        this.multiAlertServiceChanges.unsubscribe();
    };
    ClrAlertsPager.ctorParameters = function () { return [
        { type: MultiAlertService },
        { type: ClrCommonStringsService }
    ]; };
    __decorate([
        Input('clrCurrentAlert')
    ], ClrAlertsPager.prototype, "currentAlert", null);
    __decorate([
        Output('clrCurrentAlertChange')
    ], ClrAlertsPager.prototype, "currentAlertChange", void 0);
    __decorate([
        Input('clrCurrentAlertIndex')
    ], ClrAlertsPager.prototype, "currentAlertIndex", null);
    __decorate([
        Output('clrCurrentAlertIndexChange')
    ], ClrAlertsPager.prototype, "currentAlertIndexChange", void 0);
ClrAlertsPager.ɵfac = function ClrAlertsPager_Factory(t) { return new (t || ClrAlertsPager)(ɵngcc0.ɵɵdirectiveInject(MultiAlertService), ɵngcc0.ɵɵdirectiveInject(ClrCommonStringsService)); };
ClrAlertsPager.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAlertsPager, selectors: [["clr-alerts-pager"]], hostVars: 2, hostBindings: function ClrAlertsPager_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("alerts-pager", true);
    } }, inputs: { currentAlert: ["clrCurrentAlert", "currentAlert"], currentAlertIndex: ["clrCurrentAlertIndex", "currentAlertIndex"] }, outputs: { currentAlertChange: "clrCurrentAlertChange", currentAlertIndexChange: "clrCurrentAlertIndexChange" }, decls: 9, vars: 4, consts: [[1, "alerts-pager-control"], [1, "alerts-page-down"], [1, "alerts-pager-button", 3, "click"], ["shape", "caret left"], [1, "alerts-pager-text"], [1, "alerts-page-up"], ["shape", "caret right"]], template: function ClrAlertsPager_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵlistener("click", function ClrAlertsPager_Template_button_click_2_listener() { return ctx.pageDown(); });
        ɵngcc0.ɵɵelement(3, "clr-icon", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "button", 2);
        ɵngcc0.ɵɵlistener("click", function ClrAlertsPager_Template_button_click_7_listener() { return ctx.pageUp(); });
        ɵngcc0.ɵɵelement(8, "clr-icon", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("title", ctx.commonStrings.keys.previous);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate2(" ", ctx.multiAlertService.current + 1, " / ", ctx.multiAlertService.count, " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵattribute("title", ctx.commonStrings.keys.next);
    } }, directives: [ɵngcc1.ClrIconCustomTag], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAlertsPager, [{
        type: Component,
        args: [{
                selector: 'clr-alerts-pager',
                template: "<!--\n  ~ Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.\n  ~ This software is released under MIT license.\n  ~ The full license information can be found in LICENSE in the root directory of this project.\n  -->\n\n<div class=\"alerts-pager-control\">\n    <div class=\"alerts-page-down\">\n        <button class=\"alerts-pager-button\" (click)=\"pageDown()\">\n            <clr-icon shape=\"caret left\" [attr.title]=\"commonStrings.keys.previous\"></clr-icon>\n        </button>\n    </div>\n    <div class=\"alerts-pager-text\">\n        {{this.multiAlertService.current+1}} / {{this.multiAlertService.count}}\n    </div>\n    <div class=\"alerts-page-up\">\n        <button class=\"alerts-pager-button\" (click)=\"pageUp()\">\n            <clr-icon shape=\"caret right\" [attr.title]=\"commonStrings.keys.next\"></clr-icon>\n        </button>\n    </div>\n</div>\n",
                host: { '[class.alerts-pager]': 'true' }
            }]
    }], function () { return [{ type: MultiAlertService }, { type: ClrCommonStringsService }]; }, { currentAlertChange: [{
            type: Output,
            args: ['clrCurrentAlertChange']
        }], currentAlertIndexChange: [{
            type: Output,
            args: ['clrCurrentAlertIndexChange']
        }], currentAlert: [{
            type: Input,
            args: ['clrCurrentAlert']
        }], currentAlertIndex: [{
            type: Input,
            args: ['clrCurrentAlertIndex']
        }] }); })();
    return ClrAlertsPager;
}());
export { ClrAlertsPager };
