// @clr/angular/esm5/emphasis/alert/alert-item.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { AlertIconAndTypesService } from './providers/icon-and-types.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../icon/icon';

var _c0 = ["*"];
var ClrAlertItem = /** @class */ (function () {
    function ClrAlertItem(iconService) {
        this.iconService = iconService;
    }
    ClrAlertItem.ctorParameters = function () { return [
        { type: AlertIconAndTypesService }
    ]; };
ClrAlertItem.ɵfac = function ClrAlertItem_Factory(t) { return new (t || ClrAlertItem)(ɵngcc0.ɵɵdirectiveInject(AlertIconAndTypesService)); };
ClrAlertItem.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAlertItem, selectors: [["clr-alert-item"]], hostAttrs: [1, "alert-item"], ngContentSelectors: _c0, decls: 3, vars: 2, consts: [[1, "alert-icon-wrapper"], [1, "alert-icon"]], template: function ClrAlertItem_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelement(1, "clr-icon", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵattribute("shape", ctx.iconService.alertIconShape)("title", ctx.iconService.alertIconTitle);
    } }, directives: [ɵngcc1.ClrIconCustomTag], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAlertItem, [{
        type: Component,
        args: [{
                selector: 'clr-alert-item',
                template: "\n        <div class=\"alert-icon-wrapper\">\n            <clr-icon class=\"alert-icon\" \n              [attr.shape]=\"iconService.alertIconShape\" \n              [attr.title]=\"iconService.alertIconTitle\"></clr-icon>\n        </div>\n        <ng-content></ng-content>\n    ",
                host: { class: 'alert-item' }
            }]
    }], function () { return [{ type: AlertIconAndTypesService }]; }, null); })();
    return ClrAlertItem;
}());
export { ClrAlertItem };
