// @clr/angular/esm5/emphasis/alert/alert.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrDropdownModule } from '../../popover/dropdown/dropdown.module';
import { ClrAlert } from './alert';
import { ClrAlertItem } from './alert-item';
import { ClrAlerts } from './alerts';
import { ClrAlertsPager } from './alerts-pager';
import { ClrAlertText } from './alert-text';
import * as ɵngcc0 from '@angular/core';
export var CLR_ALERT_DIRECTIVES = [ClrAlert, ClrAlertItem, ClrAlerts, ClrAlertsPager, ClrAlertText];
var ClrAlertModule = /** @class */ (function () {
    function ClrAlertModule() {
    }
ClrAlertModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrAlertModule });
ClrAlertModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrAlertModule_Factory(t) { return new (t || ClrAlertModule)(); }, imports: [[CommonModule, ClrIconModule, ClrDropdownModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrAlertModule, { declarations: function () { return [ClrAlert, ClrAlertItem, ClrAlerts, ClrAlertsPager, ClrAlertText]; }, imports: function () { return [CommonModule, ClrIconModule, ClrDropdownModule]; }, exports: function () { return [ClrAlert, ClrAlertItem, ClrAlerts, ClrAlertsPager, ClrAlertText]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAlertModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrDropdownModule],
                declarations: [CLR_ALERT_DIRECTIVES],
                exports: [CLR_ALERT_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrAlertModule;
}());
export { ClrAlertModule };
