// @clr/angular/esm5/utils/drag-and-drop/drag-handle.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, OnDestroy, Optional } from '@angular/core';
import { DragHandleRegistrarService } from './providers/drag-handle-registrar.service';
import * as ɵngcc0 from '@angular/core';
var ClrDragHandle = /** @class */ (function () {
    function ClrDragHandle(el, dragHandleRegistrar) {
        this.el = el;
        this.dragHandleRegistrar = dragHandleRegistrar;
        if (!this.dragHandleRegistrar) {
            // ClrDragHandleRegistrar is provided in ClrDraggable so we expect it to be present here
            // as clrDragHandle is required to be used only inside of a clrDraggable directive.
            throw new Error('The clrDragHandle directive can only be used inside of a clrDraggable directive.');
        }
        this.dragHandleRegistrar.registerCustomHandle(this.el.nativeElement);
    }
    ClrDragHandle.prototype.ngOnDestroy = function () {
        this.dragHandleRegistrar.unregisterCustomHandle();
    };
    ClrDragHandle.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DragHandleRegistrarService, decorators: [{ type: Optional }] }
    ]; };
    ClrDragHandle = __decorate([ __param(1, Optional())
    ], ClrDragHandle);
ClrDragHandle.ɵfac = function ClrDragHandle_Factory(t) { return new (t || ClrDragHandle)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DragHandleRegistrarService, 8)); };
ClrDragHandle.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDragHandle, selectors: [["", "clrDragHandle", ""]], hostVars: 2, hostBindings: function ClrDragHandle_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("drag-handle", true);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDragHandle, [{
        type: Directive,
        args: [{ selector: '[clrDragHandle]', host: { '[class.drag-handle]': 'true' } }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: DragHandleRegistrarService, decorators: [{
                type: Optional
            }] }]; }, null); })();
    return ClrDragHandle;
}());
export { ClrDragHandle };
