// @clr/angular/esm2015/utils/focus-trap/focus-trap-tracker.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
let FocusTrapTracker = class FocusTrapTracker {
    constructor() {
        this._previousFocusTraps = [];
    }
    get current() {
        return this._current;
    }
    set current(value) {
        this._previousFocusTraps.push(this._current);
        this._current = value;
    }
    get nbFocusTrappers() {
        return this._previousFocusTraps.length;
    }
    activatePreviousTrapper() {
        this._current = this._previousFocusTraps.pop();
    }
};
FocusTrapTracker.ɵfac = function FocusTrapTracker_Factory(t) { return new (t || FocusTrapTracker)(); };
FocusTrapTracker.ɵprov = i0.ɵɵdefineInjectable({ factory: function FocusTrapTracker_Factory() { return new FocusTrapTracker(); }, token: FocusTrapTracker, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FocusTrapTracker, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
export { FocusTrapTracker };
