// @clr/angular/esm5/popover/dropdown/dropdown-trigger.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import { ClrDropdown } from './dropdown';
import { DropdownFocusHandler } from './providers/dropdown-focus-handler.service';
import * as ɵngcc0 from '@angular/core';
var ClrDropdownTrigger = /** @class */ (function () {
    function ClrDropdownTrigger(dropdown, toggleService, el, focusHandler) {
        this.toggleService = toggleService;
        this.isRootLevelToggle = true;
        // if the containing dropdown has a parent, then this is not the root level one
        if (dropdown.parent) {
            this.isRootLevelToggle = false;
        }
        focusHandler.trigger = el.nativeElement;
    }
    Object.defineProperty(ClrDropdownTrigger.prototype, "active", {
        get: function () {
            return this.toggleService.open;
        },
        enumerable: true,
        configurable: true
    });
    ClrDropdownTrigger.prototype.onDropdownTriggerClick = function (event) {
        this.toggleService.toggleWithEvent(event);
    };
    ClrDropdownTrigger.ctorParameters = function () { return [
        { type: ClrDropdown },
        { type: ClrPopoverToggleService },
        { type: ElementRef },
        { type: DropdownFocusHandler }
    ]; };
    __decorate([
        HostListener('click', ['$event'])
    ], ClrDropdownTrigger.prototype, "onDropdownTriggerClick", null);
ClrDropdownTrigger.ɵfac = function ClrDropdownTrigger_Factory(t) { return new (t || ClrDropdownTrigger)(ɵngcc0.ɵɵdirectiveInject(ClrDropdown), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DropdownFocusHandler)); };
ClrDropdownTrigger.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDropdownTrigger, selectors: [["", "clrDropdownTrigger", ""], ["", "clrDropdownToggle", ""]], hostVars: 10, hostBindings: function ClrDropdownTrigger_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrDropdownTrigger_click_HostBindingHandler($event) { return ctx.onDropdownTriggerClick($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-haspopup", "menu")("aria-expanded", ctx.active);
        ɵngcc0.ɵɵclassProp("dropdown-toggle", ctx.isRootLevelToggle)("dropdown-item", !ctx.isRootLevelToggle)("expandable", !ctx.isRootLevelToggle)("active", ctx.active);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDropdownTrigger, [{
        type: Directive,
        args: [{
                // We support both selectors for legacy reasons
                selector: '[clrDropdownTrigger],[clrDropdownToggle]',
                host: {
                    '[class.dropdown-toggle]': 'isRootLevelToggle',
                    '[class.dropdown-item]': '!isRootLevelToggle',
                    '[class.expandable]': '!isRootLevelToggle',
                    '[class.active]': 'active',
                    '[attr.aria-haspopup]': '"menu"',
                    '[attr.aria-expanded]': 'active'
                }
            }]
    }], function () { return [{ type: ClrDropdown }, { type: ClrPopoverToggleService }, { type: ɵngcc0.ElementRef }, { type: DropdownFocusHandler }]; }, { onDropdownTriggerClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
    return ClrDropdownTrigger;
}());
export { ClrDropdownTrigger };
