// @clr/angular/esm5/popover/dropdown/dropdown.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrDropdown } from './dropdown';
import { ClrDropdownItem } from './dropdown-item';
import { ClrDropdownMenu } from './dropdown-menu';
import { ClrDropdownTrigger } from './dropdown-trigger';
import * as ɵngcc0 from '@angular/core';
export var CLR_DROPDOWN_DIRECTIVES = [ClrDropdown, ClrDropdownMenu, ClrDropdownTrigger, ClrDropdownItem];
var ClrDropdownModule = /** @class */ (function () {
    function ClrDropdownModule() {
    }
ClrDropdownModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrDropdownModule });
ClrDropdownModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrDropdownModule_Factory(t) { return new (t || ClrDropdownModule)(); }, imports: [[CommonModule],
        ClrConditionalModule, ClrIconModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrDropdownModule, { declarations: function () { return [ClrDropdown, ClrDropdownMenu, ClrDropdownTrigger, ClrDropdownItem]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrDropdown, ClrDropdownMenu, ClrDropdownTrigger, ClrDropdownItem,
        ClrConditionalModule, ClrIconModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_DROPDOWN_DIRECTIVES],
                exports: [CLR_DROPDOWN_DIRECTIVES, ClrConditionalModule, ClrIconModule]
            }]
    }], function () { return []; }, null); })();
    return ClrDropdownModule;
}());
export { ClrDropdownModule };
