// @clr/angular/esm5/popover/dropdown/dropdown.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ChangeDetectorRef, Component, ElementRef, Input, OnDestroy, Optional, SkipSelf } from '@angular/core';
import { POPOVER_HOST_ANCHOR } from '../common/popover-host-anchor.token';
import { DROPDOWN_FOCUS_HANDLER_PROVIDER } from './providers/dropdown-focus-handler.service';
import { FOCUS_SERVICE_PROVIDER } from '../../utils/focus/focus.service';
import { ROOT_DROPDOWN_PROVIDER, RootDropdownService } from './providers/dropdown.service';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrDropdown = /** @class */ (function () {
    function ClrDropdown(parent, toggleService, cdr, dropdownService) {
        var _this = this;
        this.parent = parent;
        this.toggleService = toggleService;
        this.cdr = cdr;
        this.subscriptions = [];
        this.isMenuClosable = true;
        this.subscriptions.push(dropdownService.changes.subscribe(function (value) { return (_this.toggleService.open = value); }));
        this.subscriptions.push(toggleService.openChange.subscribe(function (value) { return _this.cdr.markForCheck(); }));
    }
    ClrDropdown.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
    };
    ClrDropdown.ctorParameters = function () { return [
        { type: ClrDropdown, decorators: [{ type: SkipSelf }, { type: Optional }] },
        { type: ClrPopoverToggleService },
        { type: ChangeDetectorRef },
        { type: RootDropdownService }
    ]; };
    __decorate([
        Input('clrCloseMenuOnItemClick')
    ], ClrDropdown.prototype, "isMenuClosable", void 0);
    ClrDropdown = __decorate([ __param(0, SkipSelf()),
        __param(0, Optional())
    ], ClrDropdown);
ClrDropdown.ɵfac = function ClrDropdown_Factory(t) { return new (t || ClrDropdown)(ɵngcc0.ɵɵdirectiveInject(ClrDropdown, 12), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(RootDropdownService)); };
ClrDropdown.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrDropdown, selectors: [["clr-dropdown"]], hostVars: 4, hostBindings: function ClrDropdown_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dropdown", true)("open", ctx.toggleService.open);
    } }, inputs: { isMenuClosable: ["clrCloseMenuOnItemClick", "isMenuClosable"] }, features: [ɵngcc0.ɵɵProvidersFeature([
            ROOT_DROPDOWN_PROVIDER,
            { provide: POPOVER_HOST_ANCHOR, useExisting: ElementRef },
            FOCUS_SERVICE_PROVIDER,
            ClrPopoverToggleService,
            DROPDOWN_FOCUS_HANDLER_PROVIDER,
        ])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrDropdown_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDropdown, [{
        type: Component,
        args: [{
                selector: 'clr-dropdown',
                template: '<ng-content></ng-content>',
                host: {
                    '[class.dropdown]': 'true',
                    '[class.open]': 'toggleService.open'
                },
                providers: [
                    ROOT_DROPDOWN_PROVIDER,
                    { provide: POPOVER_HOST_ANCHOR, useExisting: ElementRef },
                    FOCUS_SERVICE_PROVIDER,
                    ClrPopoverToggleService,
                    DROPDOWN_FOCUS_HANDLER_PROVIDER,
                ]
            }]
    }], function () { return [{ type: ClrDropdown, decorators: [{
                type: SkipSelf
            }, {
                type: Optional
            }] }, { type: ClrPopoverToggleService }, { type: ɵngcc0.ChangeDetectorRef }, { type: RootDropdownService }]; }, { isMenuClosable: [{
            type: Input,
            args: ['clrCloseMenuOnItemClick']
        }] }); })();
    return ClrDropdown;
}());
export { ClrDropdown };
