// @clr/angular/esm2015/utils/focus/key-focus/key-focus-item.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, PLATFORM_ID, HostBinding } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
let ClrKeyFocusItem = class ClrKeyFocusItem {
    constructor(elementRef, platformId) {
        this.elementRef = elementRef;
        this.platformId = platformId;
    }
    get nativeElement() {
        return this.elementRef.nativeElement;
    }
    focus() {
        if (isPlatformBrowser(this.platformId)) {
            this.elementRef.nativeElement.focus();
        }
    }
};
ClrKeyFocusItem.ɵfac = function ClrKeyFocusItem_Factory(t) { return new (t || ClrKeyFocusItem)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
ClrKeyFocusItem.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrKeyFocusItem, selectors: [["", "clrKeyFocusItem", ""]], hostVars: 1, hostBindings: function ClrKeyFocusItem_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("tabindex", ctx.tabIndex);
    } } });
ClrKeyFocusItem.ctorParameters = () => [
    { type: ElementRef },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
__decorate([
    HostBinding('attr.tabindex')
], ClrKeyFocusItem.prototype, "tabIndex", void 0);
ClrKeyFocusItem = __decorate([ __param(1, Inject(PLATFORM_ID))
], ClrKeyFocusItem);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrKeyFocusItem, [{
        type: Directive,
        args: [{
                selector: '[clrKeyFocusItem]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { tabIndex: [{
            type: HostBinding,
            args: ['attr.tabindex']
        }] }); })();
export { ClrKeyFocusItem };
