// @clr/angular/esm2015/popover/dropdown/providers/dropdown.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Optional, SkipSelf } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
let RootDropdownService = class RootDropdownService {
    constructor() {
        this._changes = new Subject();
    }
    get changes() {
        return this._changes.asObservable();
    }
    closeMenus() {
        this._changes.next(false);
    }
};
RootDropdownService.ɵfac = function RootDropdownService_Factory(t) { return new (t || RootDropdownService)(); };
RootDropdownService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: RootDropdownService, factory: RootDropdownService.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RootDropdownService, [{
        type: Injectable
    }], function () { return []; }, null); })();
export { RootDropdownService };
export function clrRootDropdownFactory(existing) {
    return existing || new RootDropdownService();
}
export const ROOT_DROPDOWN_PROVIDER = {
    provide: RootDropdownService,
    useFactory: clrRootDropdownFactory,
    deps: [[new Optional(), new SkipSelf(), RootDropdownService]],
};
