// @clr/angular/esm5/timeline/timeline-step-description.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrTimelineStepDescription = /** @class */ (function () {
    function ClrTimelineStepDescription() {
    }
ClrTimelineStepDescription.ɵfac = function ClrTimelineStepDescription_Factory(t) { return new (t || ClrTimelineStepDescription)(); };
ClrTimelineStepDescription.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTimelineStepDescription, selectors: [["clr-timeline-step-description"]], hostVars: 2, hostBindings: function ClrTimelineStepDescription_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-timeline-step-description", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTimelineStepDescription_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimelineStepDescription, [{
        type: Component,
        args: [{
                selector: 'clr-timeline-step-description',
                template: "\n    <ng-content></ng-content>\n  ",
                host: { '[class.clr-timeline-step-description]': 'true' }
            }]
    }], function () { return []; }, null); })();
    return ClrTimelineStepDescription;
}());
export { ClrTimelineStepDescription };
