// @clr/angular/esm5/timeline/timeline-step-title.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
/**
 * Note: Why does this component have aria-hidden attribute?
 *
 * tl;dr: we want screen readers to ignore this element when its reading out to blind users.
 *
 * In order to make a timeline step accessible to screen readers we need the title read out before the
 * icon. In order to do this, ClrTimeLine step has a ContentChild that queries for the ClrTimelineStepTitle and
 * then adds the projected text into a .clr-sr-only element that is a sibling element to the icon. See the
 * ClrTimlineStep template for the DOM structure.
 */
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrTimelineStepTitle = /** @class */ (function () {
    function ClrTimelineStepTitle() {
    }
ClrTimelineStepTitle.ɵfac = function ClrTimelineStepTitle_Factory(t) { return new (t || ClrTimelineStepTitle)(); };
ClrTimelineStepTitle.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTimelineStepTitle, selectors: [["clr-timeline-step-title"]], hostVars: 3, hostBindings: function ClrTimelineStepTitle_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-hidden", true);
        ɵngcc0.ɵɵclassProp("clr-timeline-step-title", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTimelineStepTitle_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimelineStepTitle, [{
        type: Component,
        args: [{
                selector: 'clr-timeline-step-title',
                template: "\n        <ng-content></ng-content>\n    ",
                host: { '[class.clr-timeline-step-title]': 'true', '[attr.aria-hidden]': 'true' }
            }]
    }], function () { return []; }, null); })();
    return ClrTimelineStepTitle;
}());
export { ClrTimelineStepTitle };
