// @clr/angular/esm5/timeline/timeline.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __read, __spread } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../icon/icon.module';
import { ClrSpinnerModule } from '../progress/spinner/spinner.module';
import { ClrTimeline } from './timeline';
import { ClrTimelineStep } from './timeline-step';
import { ClrTimelineStepDescription } from './timeline-step-description';
import { ClrTimelineStepHeader } from './timeline-step-header';
import { ClrTimelineStepTitle } from './timeline-step-title';
import * as ɵngcc0 from '@angular/core';
var CLR_TIMELINE_DIRECTIVES = [
    ClrTimeline,
    ClrTimelineStep,
    ClrTimelineStepDescription,
    ClrTimelineStepHeader,
    ClrTimelineStepTitle,
];
var ClrTimelineModule = /** @class */ (function () {
    function ClrTimelineModule() {
    }
ClrTimelineModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTimelineModule });
ClrTimelineModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrTimelineModule_Factory(t) { return new (t || ClrTimelineModule)(); }, imports: [[CommonModule, ClrIconModule, ClrSpinnerModule],
        ClrIconModule, ClrSpinnerModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTimelineModule, { declarations: function () { return [ClrTimeline,
        ClrTimelineStep,
        ClrTimelineStepDescription,
        ClrTimelineStepHeader,
        ClrTimelineStepTitle]; }, imports: function () { return [CommonModule, ClrIconModule, ClrSpinnerModule]; }, exports: function () { return [ClrTimeline,
        ClrTimelineStep,
        ClrTimelineStepDescription,
        ClrTimelineStepHeader,
        ClrTimelineStepTitle,
        ClrIconModule, ClrSpinnerModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimelineModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrSpinnerModule],
                exports: __spread(CLR_TIMELINE_DIRECTIVES, [ClrIconModule, ClrSpinnerModule]),
                declarations: [CLR_TIMELINE_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrTimelineModule;
}());
export { ClrTimelineModule };
