// @clr/angular/esm5/timeline/timeline-step-header.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrTimelineStepHeader = /** @class */ (function () {
    function ClrTimelineStepHeader() {
    }
ClrTimelineStepHeader.ɵfac = function ClrTimelineStepHeader_Factory(t) { return new (t || ClrTimelineStepHeader)(); };
ClrTimelineStepHeader.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTimelineStepHeader, selectors: [["clr-timeline-step-header"]], hostVars: 2, hostBindings: function ClrTimelineStepHeader_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-timeline-step-header", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTimelineStepHeader_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTimelineStepHeader, [{
        type: Component,
        args: [{
                selector: 'clr-timeline-step-header',
                template: "\n    <ng-content></ng-content>\n  ",
                host: { '[class.clr-timeline-step-header]': 'true' }
            }]
    }], function () { return []; }, null); })();
    return ClrTimelineStepHeader;
}());
export { ClrTimelineStepHeader };
