// @clr/angular/esm5/button/button-group/button-group.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrPopoverModuleNext } from '../../utils/popover/popover.module';
import { ClrButton } from './button';
import { ClrButtonGroup } from './button-group';
import * as ɵngcc0 from '@angular/core';
export var CLR_BUTTON_GROUP_DIRECTIVES = [ClrButton, ClrButtonGroup];
var ClrButtonGroupModule = /** @class */ (function () {
    function ClrButtonGroupModule() {
    }
ClrButtonGroupModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrButtonGroupModule });
ClrButtonGroupModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrButtonGroupModule_Factory(t) { return new (t || ClrButtonGroupModule)(); }, imports: [[CommonModule, ClrIconModule, ClrPopoverModuleNext]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrButtonGroupModule, { declarations: function () { return [ClrButton, ClrButtonGroup]; }, imports: function () { return [CommonModule, ClrIconModule, ClrPopoverModuleNext]; }, exports: function () { return [ClrButton, ClrButtonGroup]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrButtonGroupModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrPopoverModuleNext],
                declarations: [CLR_BUTTON_GROUP_DIRECTIVES],
                exports: [CLR_BUTTON_GROUP_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrButtonGroupModule;
}());
export { ClrButtonGroupModule };
