// @clr/angular/esm2015/utils/template-ref/template-ref.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TEMPLATE_REF_DIRECTIVES } from './index';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './template-ref-container';
let ClrTemplateRefModule = class ClrTemplateRefModule {
};
ClrTemplateRefModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTemplateRefModule });
ClrTemplateRefModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrTemplateRefModule_Factory(t) { return new (t || ClrTemplateRefModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTemplateRefModule, { declarations: function () { return [ɵngcc1.TemplateRefContainer]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ɵngcc1.TemplateRefContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTemplateRefModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [TEMPLATE_REF_DIRECTIVES],
                entryComponents: [TEMPLATE_REF_DIRECTIVES],
                exports: [TEMPLATE_REF_DIRECTIVES]
            }]
    }], null, null); })();
export { ClrTemplateRefModule };
