// @clr/angular/esm5/forms/range/range.module.js
/**
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrHostWrappingModule } from '../../utils/host-wrapping/host-wrapping.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrRange } from './range';
import { ClrRangeContainer } from './range-container';
import * as ɵngcc0 from '@angular/core';
var ClrRangeModule = /** @class */ (function () {
    function ClrRangeModule() {
    }
ClrRangeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrRangeModule });
ClrRangeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrRangeModule_Factory(t) { return new (t || ClrRangeModule)(); }, imports: [[CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrRangeModule, { declarations: function () { return [ClrRange, ClrRangeContainer]; }, imports: function () { return [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule]; }, exports: function () { return [ClrCommonFormsModule, ClrRange, ClrRangeContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrRangeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrCommonFormsModule, ClrHostWrappingModule, ClrIconModule],
                declarations: [ClrRange, ClrRangeContainer],
                exports: [ClrCommonFormsModule, ClrRange, ClrRangeContainer],
                entryComponents: [ClrRangeContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrRangeModule;
}());
export { ClrRangeModule };
