// @clr/angular/esm5/forms/datalist/datalist.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Input, Directive, Optional } from '@angular/core';
import { DatalistIdService } from './providers/datalist-id.service';
import * as ɵngcc0 from '@angular/core';
var ClrDatalist = /** @class */ (function () {
    function ClrDatalist(datalistIdService) {
        this.datalistIdService = datalistIdService;
        this.subscriptions = [];
    }
    ClrDatalist.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (!this.datalistIdService) {
            return;
        }
        this.subscriptions.push(this.datalistIdService.idChange.subscribe(function (id) { return (_this.datalistId = id); }));
    };
    Object.defineProperty(ClrDatalist.prototype, "id", {
        set: function (idValue) {
            if (!!idValue && this.datalistIdService) {
                this.datalistId = idValue;
                this.datalistIdService.id = idValue;
            }
            else if (!!idValue) {
                this.datalistId = idValue;
            }
        },
        enumerable: true,
        configurable: true
    });
    ClrDatalist.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
    };
    ClrDatalist.ctorParameters = function () { return [
        { type: DatalistIdService, decorators: [{ type: Optional }] }
    ]; };
    __decorate([
        Input()
    ], ClrDatalist.prototype, "id", null);
    ClrDatalist = __decorate([ __param(0, Optional())
    ], ClrDatalist);
ClrDatalist.ɵfac = function ClrDatalist_Factory(t) { return new (t || ClrDatalist)(ɵngcc0.ɵɵdirectiveInject(DatalistIdService, 8)); };
ClrDatalist.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDatalist, selectors: [["datalist"]], hostVars: 1, hostBindings: function ClrDatalist_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("id", ctx.datalistId);
    } }, inputs: { id: "id" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDatalist, [{
        type: Directive,
        args: [{
                selector: 'datalist',
                host: {
                    '[id]': 'datalistId'
                }
            }]
    }], function () { return [{ type: DatalistIdService, decorators: [{
                type: Optional
            }] }]; }, { id: [{
            type: Input
        }] }); })();
    return ClrDatalist;
}());
export { ClrDatalist };
