// @clr/angular/esm5/forms/datalist/datalist.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrDatalistInput } from './datalist-input';
import { ClrInputModule } from '../input/input.module';
import { ClrDatalistContainer } from './datalist-container';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrDatalist } from './datalist';
import * as ɵngcc0 from '@angular/core';
var ClrDatalistModule = /** @class */ (function () {
    function ClrDatalistModule() {
    }
ClrDatalistModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrDatalistModule });
ClrDatalistModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrDatalistModule_Factory(t) { return new (t || ClrDatalistModule)(); }, imports: [[CommonModule, ClrInputModule, ClrIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrDatalistModule, { declarations: function () { return [ClrDatalist, ClrDatalistInput, ClrDatalistContainer]; }, imports: function () { return [CommonModule, ClrInputModule, ClrIconModule]; }, exports: function () { return [ClrDatalist, ClrDatalistInput, ClrDatalistContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDatalistModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrInputModule, ClrIconModule],
                declarations: [ClrDatalist, ClrDatalistInput, ClrDatalistContainer],
                exports: [ClrDatalist, ClrDatalistInput, ClrDatalistContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrDatalistModule;
}());
export { ClrDatalistModule };
