// @clr/angular/esm5/forms/datalist/datalist-input.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __extends, __param } from "tslib";
import { Directive, ViewContainerRef, Injector, Self, Optional, Renderer2, ElementRef, HostListener, AfterContentInit, } from '@angular/core';
import { WrappedFormControl } from '../common/wrapped-control';
import { ClrDatalistContainer } from './datalist-container';
import { NgControl } from '@angular/forms';
import { FocusService } from '../common/providers/focus.service';
import { DatalistIdService } from './providers/datalist-id.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
var ClrDatalistInput = /** @class */ (function (_super) {
    __extends(ClrDatalistInput, _super);
    function ClrDatalistInput(focusService, vcr, injector, control, renderer, el, datalistIdService) {
        var _this = _super.call(this, vcr, ClrDatalistContainer, injector, control, renderer, el) || this;
        _this.focusService = focusService;
        _this.datalistIdService = datalistIdService;
        if (!_this.focusService) {
            throw new Error('clrDatalist requires being wrapped in <clr-datalist-container>');
        }
        return _this;
    }
    ClrDatalistInput.prototype.ngAfterContentInit = function () {
        var _this = this;
        // Subscriptions is inherited from WrappedFormControl, unsubscribe is handled there
        this.subscriptions.push(this.datalistIdService.idChange.subscribe(function (id) { return (_this.listValue = id); }));
    };
    ClrDatalistInput.prototype.triggerFocus = function () {
        if (this.focusService) {
            this.focusService.focused = true;
        }
    };
    ClrDatalistInput.prototype.triggerValidation = function () {
        _super.prototype.triggerValidation.call(this);
        if (this.focusService) {
            this.focusService.focused = false;
        }
    };
    ClrDatalistInput.ctorParameters = function () { return [
        { type: FocusService, decorators: [{ type: Optional }] },
        { type: ViewContainerRef },
        { type: Injector },
        { type: NgControl, decorators: [{ type: Self }, { type: Optional }] },
        { type: Renderer2 },
        { type: ElementRef },
        { type: DatalistIdService }
    ]; };
    __decorate([
        HostListener('focus')
    ], ClrDatalistInput.prototype, "triggerFocus", null);
    __decorate([
        HostListener('blur')
    ], ClrDatalistInput.prototype, "triggerValidation", null);
    ClrDatalistInput = __decorate([ __param(0, Optional()),
        __param(3, Self()),
        __param(3, Optional())
    ], ClrDatalistInput);
ClrDatalistInput.ɵfac = function ClrDatalistInput_Factory(t) { return new (t || ClrDatalistInput)(ɵngcc0.ɵɵdirectiveInject(FocusService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgControl, 10), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DatalistIdService)); };
ClrDatalistInput.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDatalistInput, selectors: [["", "clrDatalistInput", ""]], hostVars: 3, hostBindings: function ClrDatalistInput_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("focus", function ClrDatalistInput_focus_HostBindingHandler() { return ctx.triggerFocus(); })("blur", function ClrDatalistInput_blur_HostBindingHandler() { return ctx.triggerValidation(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("list", ctx.listValue);
        ɵngcc0.ɵɵclassProp("clr-input", true);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDatalistInput, [{
        type: Directive,
        args: [{
                selector: '[clrDatalistInput]',
                host: {
                    '[class.clr-input]': 'true',
                    '[attr.list]': 'listValue'
                }
            }]
    }], function () { return [{ type: FocusService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.Injector }, { type: ɵngcc1.NgControl, decorators: [{
                type: Self
            }, {
                type: Optional
            }] }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }, { type: DatalistIdService }]; }, { triggerFocus: [{
            type: HostListener,
            args: ['focus']
        }], triggerValidation: [{
            type: HostListener,
            args: ['blur']
        }] }); })();
    return ClrDatalistInput;
}(WrappedFormControl));
export { ClrDatalistInput };
