// @clr/angular/esm5/forms/datepicker/providers/date-form-control.service.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
var DateFormControlService = /** @class */ (function () {
    function DateFormControlService() {
        this._touchedChange = new Subject();
        this._dirtyChange = new Subject();
    }
    Object.defineProperty(DateFormControlService.prototype, "touchedChange", {
        get: function () {
            return this._touchedChange.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DateFormControlService.prototype, "dirtyChange", {
        get: function () {
            return this._dirtyChange.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    DateFormControlService.prototype.markAsTouched = function () {
        this._touchedChange.next();
    };
    DateFormControlService.prototype.markAsDirty = function () {
        this._dirtyChange.next();
    };
    // friendly wrapper
    DateFormControlService.prototype.setDisabled = function (state) {
        this.disabled = state;
    };
DateFormControlService.ɵfac = function DateFormControlService_Factory(t) { return new (t || DateFormControlService)(); };
DateFormControlService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DateFormControlService, factory: function (t) { return DateFormControlService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DateFormControlService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return DateFormControlService;
}());
export { DateFormControlService };
