// @clr/angular/esm5/forms/datepicker/providers/view-manager.service.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ClrPopoverPositions } from '../../../utils/popover/enums/positions.enum';
/**
 * This service manages which view is visible in the datepicker popover.
 */
import * as ɵngcc0 from '@angular/core';
var ViewManagerService = /** @class */ (function () {
    function ViewManagerService() {
        this._currentView = "DAYVIEW" /* DAYVIEW */;
        this.position = ClrPopoverPositions['bottom-left'];
    }
    Object.defineProperty(ViewManagerService.prototype, "isDayView", {
        get: function () {
            return this._currentView === "DAYVIEW" /* DAYVIEW */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewManagerService.prototype, "isYearView", {
        get: function () {
            return this._currentView === "YEARVIEW" /* YEARVIEW */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewManagerService.prototype, "isMonthView", {
        get: function () {
            return this._currentView === "MONTHVIEW" /* MONTHVIEW */;
        },
        enumerable: true,
        configurable: true
    });
    ViewManagerService.prototype.changeToMonthView = function () {
        this._currentView = "MONTHVIEW" /* MONTHVIEW */;
    };
    ViewManagerService.prototype.changeToYearView = function () {
        this._currentView = "YEARVIEW" /* YEARVIEW */;
    };
    ViewManagerService.prototype.changeToDayView = function () {
        this._currentView = "DAYVIEW" /* DAYVIEW */;
    };
ViewManagerService.ɵfac = function ViewManagerService_Factory(t) { return new (t || ViewManagerService)(); };
ViewManagerService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ViewManagerService, factory: function (t) { return ViewManagerService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ViewManagerService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return ViewManagerService;
}());
export { ViewManagerService };
