// @clr/angular/esm5/utils/focus-trap/focus-trap.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FocusTrapDirective } from './focus-trap.directive';
import * as ɵngcc0 from '@angular/core';
export var FOCUS_TRAP_DIRECTIVES = [FocusTrapDirective];
var ClrFocusTrapModule = /** @class */ (function () {
    function ClrFocusTrapModule() {
    }
ClrFocusTrapModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrFocusTrapModule });
ClrFocusTrapModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrFocusTrapModule_Factory(t) { return new (t || ClrFocusTrapModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrFocusTrapModule, { declarations: function () { return [FocusTrapDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [FocusTrapDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrFocusTrapModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [FOCUS_TRAP_DIRECTIVES],
                exports: [FOCUS_TRAP_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrFocusTrapModule;
}());
export { ClrFocusTrapModule };
