// @clr/angular/esm2015/utils/drag-and-drop/if-dragged.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, OnDestroy, Optional, SkipSelf, TemplateRef, ViewContainerRef } from '@angular/core';
import { DragEventListenerService } from './providers/drag-event-listener.service';
// This structural directive will be used mainly together with `clr-draggable-ghost` directive inside of clrDraggable
// directive. The directive is responsible for instantiating `clr-draggable-ghost` directive only during dragging so
// that Angular Change Detection is prevented from running if a component or directive is placed inside of the
// `clr-draggable-ghost` directive.
import * as ɵngcc0 from '@angular/core';
let ClrIfDragged = class ClrIfDragged {
    constructor(template, container, dragEventListener) {
        this.template = template;
        this.container = container;
        this.dragEventListener = dragEventListener;
        this.subscriptions = [];
        if (!this.dragEventListener || !this.container) {
            throw new Error('The *clrIfDragged directive can only be used inside of a clrDraggable directive.');
        }
        this.subscriptions.push(this.dragEventListener.dragStarted.subscribe((event) => {
            this.container.createEmbeddedView(this.template);
        }));
        this.subscriptions.push(this.dragEventListener.dragEnded.subscribe((event) => {
            this.container.clear();
        }));
    }
    ngOnDestroy() {
        this.subscriptions.forEach((sub) => sub.unsubscribe());
    }
};
ClrIfDragged.ɵfac = function ClrIfDragged_Factory(t) { return new (t || ClrIfDragged)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef, 12), ɵngcc0.ɵɵdirectiveInject(DragEventListenerService, 8)); };
ClrIfDragged.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrIfDragged, selectors: [["", "clrIfDragged", ""]] });
ClrIfDragged.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef, decorators: [{ type: Optional }, { type: SkipSelf }] },
    { type: DragEventListenerService, decorators: [{ type: Optional }] }
];
ClrIfDragged = __decorate([ __param(1, Optional()),
    __param(1, SkipSelf()),
    __param(2, Optional())
], ClrIfDragged);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrIfDragged, [{
        type: Directive,
        args: [{ selector: '[clrIfDragged]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }, { type: DragEventListenerService, decorators: [{
                type: Optional
            }] }]; }, null); })();
export { ClrIfDragged };
