// @clr/angular/esm2015/popover/tooltip/tooltip-trigger.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Directive, HostListener } from '@angular/core';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import { TooltipIdService } from './providers/tooltip-id.service';
import * as ɵngcc0 from '@angular/core';
let ClrTooltipTrigger = class ClrTooltipTrigger {
    constructor(toggleService, tooltipIdService) {
        this.toggleService = toggleService;
        this.tooltipIdService = tooltipIdService;
        this.subs = [];
        // The aria-described by comes from the id of content. It
        this.subs.push(this.tooltipIdService.id.subscribe(tooltipId => (this.ariaDescribedBy = tooltipId)));
    }
    showTooltip() {
        this.toggleService.open = true;
    }
    hideTooltip() {
        this.toggleService.open = false;
    }
    ngOnDestroy() {
        this.subs.forEach(sub => sub.unsubscribe());
    }
};
ClrTooltipTrigger.ɵfac = function ClrTooltipTrigger_Factory(t) { return new (t || ClrTooltipTrigger)(ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(TooltipIdService)); };
ClrTooltipTrigger.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrTooltipTrigger, selectors: [["", "clrTooltipTrigger", ""]], hostAttrs: ["tabindex", "0"], hostVars: 4, hostBindings: function ClrTooltipTrigger_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mouseenter", function ClrTooltipTrigger_mouseenter_HostBindingHandler() { return ctx.showTooltip(); })("focus", function ClrTooltipTrigger_focus_HostBindingHandler() { return ctx.showTooltip(); })("mouseleave", function ClrTooltipTrigger_mouseleave_HostBindingHandler() { return ctx.hideTooltip(); })("blur", function ClrTooltipTrigger_blur_HostBindingHandler() { return ctx.hideTooltip(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-describedby", ctx.ariaDescribedBy)("role", "button");
        ɵngcc0.ɵɵclassProp("tooltip-trigger", true);
    } } });
ClrTooltipTrigger.ctorParameters = () => [
    { type: ClrPopoverToggleService },
    { type: TooltipIdService }
];
__decorate([
    HostListener('mouseenter'),
    HostListener('focus')
], ClrTooltipTrigger.prototype, "showTooltip", null);
__decorate([
    HostListener('mouseleave'),
    HostListener('blur')
], ClrTooltipTrigger.prototype, "hideTooltip", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTooltipTrigger, [{
        type: Directive,
        args: [{
                selector: '[clrTooltipTrigger]',
                host: {
                    tabindex: '0',
                    '[class.tooltip-trigger]': 'true',
                    '[attr.aria-describedby]': 'ariaDescribedBy',
                    '[attr.role]': '"button"'
                }
            }]
    }], function () { return [{ type: ClrPopoverToggleService }, { type: TooltipIdService }]; }, { showTooltip: [{
            type: HostListener,
            args: ['mouseenter']
        }, {
            type: HostListener,
            args: ['focus']
        }], hideTooltip: [{
            type: HostListener,
            args: ['mouseleave']
        }, {
            type: HostListener,
            args: ['blur']
        }] }); })();
export { ClrTooltipTrigger };
