// @clr/angular/esm5/modal/modal.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../icon/icon.module';
import { ClrFocusTrapModule } from '../utils/focus-trap/focus-trap.module';
import { ClrModal } from './modal';
import { ClrModalBody } from './modal-body';
import { ClrFocusOnViewInitModule } from '../utils/focus/focus-on-view-init/focus-on-view-init.module';
import * as ɵngcc0 from '@angular/core';
export var CLR_MODAL_DIRECTIVES = [ClrModal, ClrModalBody];
var ClrModalModule = /** @class */ (function () {
    function ClrModalModule() {
    }
ClrModalModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrModalModule });
ClrModalModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrModalModule_Factory(t) { return new (t || ClrModalModule)(); }, imports: [[CommonModule, ClrIconModule, ClrFocusTrapModule, ClrFocusOnViewInitModule],
        ClrIconModule, ClrFocusOnViewInitModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrModalModule, { declarations: function () { return [ClrModal, ClrModalBody]; }, imports: function () { return [CommonModule, ClrIconModule, ClrFocusTrapModule, ClrFocusOnViewInitModule]; }, exports: function () { return [ClrModal, ClrModalBody,
        ClrIconModule, ClrFocusOnViewInitModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrModalModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrFocusTrapModule, ClrFocusOnViewInitModule],
                declarations: [CLR_MODAL_DIRECTIVES],
                exports: [CLR_MODAL_DIRECTIVES, ClrIconModule, ClrFocusOnViewInitModule]
            }]
    }], function () { return []; }, null); })();
    return ClrModalModule;
}());
export { ClrModalModule };
