// @clr/angular/esm5/utils/template-ref/template-ref-container.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, TemplateRef, ViewChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

function TemplateRefContainer_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c0 = ["*"];
var TemplateRefContainer = /** @class */ (function () {
    function TemplateRefContainer() {
    }
    __decorate([
        ViewChild(TemplateRef)
    ], TemplateRefContainer.prototype, "template", void 0);
TemplateRefContainer.ɵfac = function TemplateRefContainer_Factory(t) { return new (t || TemplateRefContainer)(); };
TemplateRefContainer.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TemplateRefContainer, selectors: [["ng-component"]], viewQuery: function TemplateRefContainer_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function TemplateRefContainer_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, TemplateRefContainer_ng_template_0_Template, 1, 0, "ng-template");
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TemplateRefContainer, [{
        type: Component,
        args: [{
                template: "\n      <ng-template>\n        <ng-content></ng-content>\n      </ng-template>\n    "
            }]
    }], function () { return []; }, { template: [{
            type: ViewChild,
            args: [TemplateRef]
        }] }); })();
    return TemplateRefContainer;
}());
export { TemplateRefContainer };
