// @clr/angular/esm5/forms/datepicker/datepicker-view-manager.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { DatepickerFocusService } from './providers/datepicker-focus.service';
import { ViewManagerService } from './providers/view-manager.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from './monthpicker';
import * as ɵngcc3 from './yearpicker';
import * as ɵngcc4 from './daypicker';

function ClrDatepickerViewManager_clr_monthpicker_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "clr-monthpicker");
} }
function ClrDatepickerViewManager_clr_yearpicker_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "clr-yearpicker");
} }
function ClrDatepickerViewManager_clr_daypicker_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "clr-daypicker");
} }
var ClrDatepickerViewManager = /** @class */ (function () {
    function ClrDatepickerViewManager(viewManagerService) {
        this.viewManagerService = viewManagerService;
    }
    Object.defineProperty(ClrDatepickerViewManager.prototype, "isMonthView", {
        /**
         * Returns if the current view is the monthpicker.
         */
        get: function () {
            return this.viewManagerService.isMonthView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrDatepickerViewManager.prototype, "isYearView", {
        /**
         * Returns if the current view is the yearpicker.
         */
        get: function () {
            return this.viewManagerService.isYearView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrDatepickerViewManager.prototype, "isDayView", {
        /**
         * Returns if the current view is the daypicker.
         */
        get: function () {
            return this.viewManagerService.isDayView;
        },
        enumerable: true,
        configurable: true
    });
    ClrDatepickerViewManager.ctorParameters = function () { return [
        { type: ViewManagerService }
    ]; };
ClrDatepickerViewManager.ɵfac = function ClrDatepickerViewManager_Factory(t) { return new (t || ClrDatepickerViewManager)(ɵngcc0.ɵɵdirectiveInject(ViewManagerService)); };
ClrDatepickerViewManager.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrDatepickerViewManager, selectors: [["clr-datepicker-view-manager"]], hostVars: 3, hostBindings: function ClrDatepickerViewManager_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-modal", true);
        ɵngcc0.ɵɵclassProp("datepicker", true);
    } }, features: [ɵngcc0.ɵɵProvidersFeature([DatepickerFocusService])], decls: 3, vars: 3, consts: [[4, "ngIf"]], template: function ClrDatepickerViewManager_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ClrDatepickerViewManager_clr_monthpicker_0_Template, 1, 0, "clr-monthpicker", 0);
        ɵngcc0.ɵɵtemplate(1, ClrDatepickerViewManager_clr_yearpicker_1_Template, 1, 0, "clr-yearpicker", 0);
        ɵngcc0.ɵɵtemplate(2, ClrDatepickerViewManager_clr_daypicker_2_Template, 1, 0, "clr-daypicker", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isMonthView);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isYearView);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDayView);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.ClrMonthpicker, ɵngcc3.ClrYearpicker, ɵngcc4.ClrDaypicker], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDatepickerViewManager, [{
        type: Component,
        args: [{
                selector: 'clr-datepicker-view-manager',
                template: "<!--\n* Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.\n* This software is released under MIT license.\n* The full license information can be found in LICENSE in the root directory of this project.\n-->\n\n<clr-monthpicker *ngIf=\"isMonthView\"></clr-monthpicker>\n<clr-yearpicker *ngIf=\"isYearView\"></clr-yearpicker>\n<clr-daypicker *ngIf=\"isDayView\"></clr-daypicker>\n",
                providers: [DatepickerFocusService],
                host: {
                    '[class.datepicker]': 'true',
                    '[attr.aria-modal]': 'true'
                }
            }]
    }], function () { return [{ type: ViewManagerService }]; }, null); })();
    return ClrDatepickerViewManager;
}());
export { ClrDatepickerViewManager };
