// @clr/angular/esm5/forms/datepicker/date-input.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __extends, __param, __read } from "tslib";
import { isPlatformBrowser } from '@angular/common';
import { AfterViewInit, Directive, ElementRef, EventEmitter, HostBinding, HostListener, Inject, Injector, Input, OnDestroy, OnInit, Optional, Output, PLATFORM_ID, Renderer2, Self, ViewContainerRef, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { filter, switchMap } from 'rxjs/operators';
import { of } from 'rxjs';
import { FocusService } from '../common/providers/focus.service';
import { WrappedFormControl } from '../common/wrapped-control';
import { ClrDateContainer } from './date-container';
import { DayModel } from './model/day.model';
import { DateFormControlService } from './providers/date-form-control.service';
import { DateIOService } from './providers/date-io.service';
import { DateNavigationService } from './providers/date-navigation.service';
import { DatepickerEnabledService } from './providers/datepicker-enabled.service';
import { DatepickerFocusService } from './providers/datepicker-focus.service';
import { datesAreEqual } from './utils/date-utils';
import { isBooleanAttributeSet } from '../../utils/component/is-boolean-attribute-set';
// There are four ways the datepicker value is set
// 1. Value set by user typing into text input as a string ex: '01/28/2015'
// 2. Value set explicitly by Angular Forms APIs as a string ex: '01/28/2015'
// 3. Value set by user via datepicker UI as a Date Object
// 4. Value set via `clrDate` input as a Date Object
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
var ClrDateInput = /** @class */ (function (_super) {
    __extends(ClrDateInput, _super);
    function ClrDateInput(viewContainerRef, injector, el, renderer, control, container, dateIOService, dateNavigationService, datepickerEnabledService, dateFormControlService, platformId, focusService, datepickerFocusService) {
        var _this = _super.call(this, viewContainerRef, ClrDateContainer, injector, control, renderer, el) || this;
        _this.el = el;
        _this.renderer = renderer;
        _this.control = control;
        _this.container = container;
        _this.dateIOService = dateIOService;
        _this.dateNavigationService = dateNavigationService;
        _this.datepickerEnabledService = datepickerEnabledService;
        _this.dateFormControlService = dateFormControlService;
        _this.platformId = platformId;
        _this.focusService = focusService;
        _this.datepickerFocusService = datepickerFocusService;
        _this.dateChange = new EventEmitter(false);
        _this.index = 1;
        return _this;
    }
    Object.defineProperty(ClrDateInput.prototype, "date", {
        set: function (date) {
            if (this.previousDateChange !== date) {
                this.updateDate(this.getValidDateValueFromDate(date));
            }
            if (!this.initialClrDateInputValue) {
                this.initialClrDateInputValue = date;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrDateInput.prototype, "min", {
        set: function (dateString) {
            this.dateIOService.setMinDate(dateString);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrDateInput.prototype, "max", {
        set: function (dateString) {
            this.dateIOService.setMaxDate(dateString);
        },
        enumerable: true,
        configurable: true
    });
    ClrDateInput.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
        this.populateServicesFromContainerComponent();
        this.subscriptions.push(this.listenForUserSelectedDayChanges(), this.listenForControlValueChanges(), this.listenForTouchChanges(), this.listenForDirtyChanges(), this.listenForInputRefocus());
    };
    ClrDateInput.prototype.ngAfterViewInit = function () {
        // I don't know why I have to do this but after using the new HostWrapping Module I have to delay the processing
        // of the initial Input set by the user to here. If I do not 2 issues occur:
        // 1. The Input setter is called before ngOnInit. ngOnInit initializes the services without which the setter fails.
        // 2. The Renderer doesn't work before ngAfterViewInit (It used to before the new HostWrapping Module for some reason).
        // I need the renderer to set the value property on the input to make sure that if the user has supplied a Date
        // input object, we reflect it with the right date on the input field using the IO service. I am not sure if
        // these are major issues or not but just noting them down here.
        this.processInitialInputs();
    };
    ClrDateInput.prototype.setFocusStates = function () {
        this.setFocus(true);
    };
    ClrDateInput.prototype.triggerValidation = function () {
        _super.prototype.triggerValidation.call(this);
        this.setFocus(false);
    };
    Object.defineProperty(ClrDateInput.prototype, "placeholderText", {
        get: function () {
            return this.placeholder ? this.placeholder : this.dateIOService.placeholderText;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrDateInput.prototype, "inputType", {
        get: function () {
            return isPlatformBrowser(this.platformId) && this.usingNativeDatepicker() ? 'date' : 'text';
        },
        enumerable: true,
        configurable: true
    });
    ClrDateInput.prototype.onValueChange = function (target) {
        var validDateValue = this.dateIOService.getDateValueFromDateString(target.value);
        if (this.usingClarityDatepicker() && validDateValue) {
            this.updateDate(validDateValue, true);
        }
        else if (this.usingNativeDatepicker()) {
            var _a = __read(target.value.split('-'), 3), year = _a[0], month = _a[1], day = _a[2];
            this.updateDate(new Date(+year, +month - 1, +day), true);
        }
        else {
            this.emitDateOutput(null);
        }
    };
    Object.defineProperty(ClrDateInput.prototype, "disabled", {
        get: function () {
            if (this.dateFormControlService) {
                return this.dateFormControlService.disabled;
            }
            return null;
        },
        set: function (value) {
            if (this.dateFormControlService) {
                this.dateFormControlService.setDisabled(isBooleanAttributeSet(value));
            }
        },
        enumerable: true,
        configurable: true
    });
    ClrDateInput.prototype.usingClarityDatepicker = function () {
        return this.datepickerEnabledService.isEnabled;
    };
    ClrDateInput.prototype.usingNativeDatepicker = function () {
        return !this.datepickerEnabledService.isEnabled;
    };
    ClrDateInput.prototype.setFocus = function (focus) {
        if (this.focusService) {
            this.focusService.focused = focus;
        }
    };
    ClrDateInput.prototype.populateServicesFromContainerComponent = function () {
        if (!this.container) {
            this.dateIOService = this.getProviderFromContainer(DateIOService);
            this.dateNavigationService = this.getProviderFromContainer(DateNavigationService);
            this.datepickerEnabledService = this.getProviderFromContainer(DatepickerEnabledService);
            this.dateFormControlService = this.getProviderFromContainer(DateFormControlService);
        }
    };
    ClrDateInput.prototype.processInitialInputs = function () {
        if (this.datepickerHasFormControl()) {
            this.updateDate(this.dateIOService.getDateValueFromDateString(this.control.value));
        }
        else {
            this.updateDate(this.initialClrDateInputValue);
        }
    };
    ClrDateInput.prototype.updateDate = function (value, setByUserInteraction) {
        if (setByUserInteraction === void 0) { setByUserInteraction = false; }
        var date = this.getValidDateValueFromDate(value);
        if (setByUserInteraction) {
            this.emitDateOutput(date);
        }
        else {
            this.previousDateChange = date;
        }
        if (this.dateNavigationService) {
            this.dateNavigationService.selectedDay = date
                ? new DayModel(date.getFullYear(), date.getMonth(), date.getDate())
                : null;
        }
        this.updateInput(date);
    };
    ClrDateInput.prototype.updateInput = function (date) {
        if (date) {
            var dateString = this.dateIOService.toLocaleDisplayFormatString(date);
            if (this.usingNativeDatepicker()) {
                // valueAsDate expects UTC, date from input is time-zoned
                date.setMinutes(date.getMinutes() - date.getTimezoneOffset());
                this.renderer.setProperty(this.el.nativeElement, 'valueAsDate', date);
            }
            else if (this.datepickerHasFormControl() && dateString !== this.control.value) {
                this.control.control.setValue(dateString);
            }
            else {
                this.renderer.setProperty(this.el.nativeElement, 'value', dateString);
            }
        }
        else {
            this.renderer.setProperty(this.el.nativeElement, 'value', '');
        }
    };
    ClrDateInput.prototype.getValidDateValueFromDate = function (date) {
        if (this.dateIOService) {
            var dateString = this.dateIOService.toLocaleDisplayFormatString(date);
            return this.dateIOService.getDateValueFromDateString(dateString);
        }
        else {
            return null;
        }
    };
    ClrDateInput.prototype.emitDateOutput = function (date) {
        if (!datesAreEqual(date, this.previousDateChange)) {
            this.dateChange.emit(date);
            this.previousDateChange = date;
        }
        else if (!date && this.previousDateChange) {
            this.dateChange.emit(null);
            this.previousDateChange = null;
        }
    };
    ClrDateInput.prototype.datepickerHasFormControl = function () {
        return !!this.control;
    };
    ClrDateInput.prototype.listenForControlValueChanges = function () {
        var _this = this;
        return of(this.datepickerHasFormControl())
            .pipe(filter(function (hasControl) { return hasControl; }), switchMap(function () { return _this.control.valueChanges; }), 
        // only update date value if not being set by user
        filter(function () { return !_this.datepickerFocusService.elementIsFocused(_this.el.nativeElement); }))
            .subscribe(function (value) { return _this.updateDate(_this.dateIOService.getDateValueFromDateString(value)); });
    };
    ClrDateInput.prototype.listenForUserSelectedDayChanges = function () {
        var _this = this;
        return this.dateNavigationService.selectedDayChange.subscribe(function (dayModel) { return _this.updateDate(dayModel.toDate(), true); });
    };
    ClrDateInput.prototype.listenForTouchChanges = function () {
        var _this = this;
        return this.dateFormControlService.touchedChange
            .pipe(filter(function () { return _this.datepickerHasFormControl(); }))
            .subscribe(function () { return _this.control.control.markAsTouched(); });
    };
    ClrDateInput.prototype.listenForDirtyChanges = function () {
        var _this = this;
        return this.dateFormControlService.dirtyChange
            .pipe(filter(function () { return _this.datepickerHasFormControl(); }))
            .subscribe(function () { return _this.control.control.markAsDirty(); });
    };
    ClrDateInput.prototype.listenForInputRefocus = function () {
        var _this = this;
        return this.dateNavigationService.selectedDayChange
            .pipe(filter(function (date) { return !!date; }))
            .subscribe(function (v) { return _this.datepickerFocusService.focusInput(_this.el.nativeElement); });
    };
    ClrDateInput.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Injector },
        { type: ElementRef },
        { type: Renderer2 },
        { type: NgControl, decorators: [{ type: Self }, { type: Optional }] },
        { type: ClrDateContainer, decorators: [{ type: Optional }] },
        { type: DateIOService, decorators: [{ type: Optional }] },
        { type: DateNavigationService, decorators: [{ type: Optional }] },
        { type: DatepickerEnabledService, decorators: [{ type: Optional }] },
        { type: DateFormControlService, decorators: [{ type: Optional }] },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: FocusService, decorators: [{ type: Optional }] },
        { type: DatepickerFocusService }
    ]; };
    __decorate([
        Input()
    ], ClrDateInput.prototype, "placeholder", void 0);
    __decorate([
        Output('clrDateChange')
    ], ClrDateInput.prototype, "dateChange", void 0);
    __decorate([
        Input('clrDate')
    ], ClrDateInput.prototype, "date", null);
    __decorate([
        Input()
    ], ClrDateInput.prototype, "min", null);
    __decorate([
        Input()
    ], ClrDateInput.prototype, "max", null);
    __decorate([
        HostListener('focus')
    ], ClrDateInput.prototype, "setFocusStates", null);
    __decorate([
        HostListener('blur')
    ], ClrDateInput.prototype, "triggerValidation", null);
    __decorate([
        HostBinding('attr.placeholder')
    ], ClrDateInput.prototype, "placeholderText", null);
    __decorate([
        HostBinding('attr.type')
    ], ClrDateInput.prototype, "inputType", null);
    __decorate([
        HostListener('change', ['$event.target'])
    ], ClrDateInput.prototype, "onValueChange", null);
    __decorate([
        HostBinding('disabled'),
        Input('disabled')
    ], ClrDateInput.prototype, "disabled", null);
    ClrDateInput = __decorate([ __param(4, Self()),
        __param(4, Optional()),
        __param(5, Optional()),
        __param(6, Optional()),
        __param(7, Optional()),
        __param(8, Optional()),
        __param(9, Optional()),
        __param(10, Inject(PLATFORM_ID)),
        __param(11, Optional())
    ], ClrDateInput);
ClrDateInput.ɵfac = function ClrDateInput_Factory(t) { return new (t || ClrDateInput)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgControl, 10), ɵngcc0.ɵɵdirectiveInject(ClrDateContainer, 8), ɵngcc0.ɵɵdirectiveInject(DateIOService, 8), ɵngcc0.ɵɵdirectiveInject(DateNavigationService, 8), ɵngcc0.ɵɵdirectiveInject(DatepickerEnabledService, 8), ɵngcc0.ɵɵdirectiveInject(DateFormControlService, 8), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID), ɵngcc0.ɵɵdirectiveInject(FocusService, 8), ɵngcc0.ɵɵdirectiveInject(DatepickerFocusService)); };
ClrDateInput.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrDateInput, selectors: [["", "clrDate", ""]], hostVars: 5, hostBindings: function ClrDateInput_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("focus", function ClrDateInput_focus_HostBindingHandler() { return ctx.setFocusStates(); })("blur", function ClrDateInput_blur_HostBindingHandler() { return ctx.triggerValidation(); })("change", function ClrDateInput_change_HostBindingHandler($event) { return ctx.onValueChange($event.target); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("disabled", ctx.disabled);
        ɵngcc0.ɵɵattribute("placeholder", ctx.placeholderText)("type", ctx.inputType);
        ɵngcc0.ɵɵclassProp("clr-input", true);
    } }, inputs: { date: ["clrDate", "date"], min: "min", max: "max", disabled: "disabled", placeholder: "placeholder" }, outputs: { dateChange: "clrDateChange" }, features: [ɵngcc0.ɵɵProvidersFeature([DatepickerFocusService]), ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDateInput, [{
        type: Directive,
        args: [{
                selector: '[clrDate]',
                host: {
                    '[class.clr-input]': 'true'
                },
                providers: [DatepickerFocusService]
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.Injector }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc1.NgControl, decorators: [{
                type: Self
            }, {
                type: Optional
            }] }, { type: ClrDateContainer, decorators: [{
                type: Optional
            }] }, { type: DateIOService, decorators: [{
                type: Optional
            }] }, { type: DateNavigationService, decorators: [{
                type: Optional
            }] }, { type: DatepickerEnabledService, decorators: [{
                type: Optional
            }] }, { type: DateFormControlService, decorators: [{
                type: Optional
            }] }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }, { type: FocusService, decorators: [{
                type: Optional
            }] }, { type: DatepickerFocusService }]; }, { date: [{
            type: Input,
            args: ['clrDate']
        }], min: [{
            type: Input
        }], max: [{
            type: Input
        }], setFocusStates: [{
            type: HostListener,
            args: ['focus']
        }], triggerValidation: [{
            type: HostListener,
            args: ['blur']
        }], placeholderText: [{
            type: HostBinding,
            args: ['attr.placeholder']
        }], inputType: [{
            type: HostBinding,
            args: ['attr.type']
        }], onValueChange: [{
            type: HostListener,
            args: ['change', ['$event.target']]
        }], disabled: [{
            type: HostBinding,
            args: ['disabled']
        }, {
            type: Input,
            args: ['disabled']
        }], placeholder: [{
            type: Input
        }], dateChange: [{
            type: Output,
            args: ['clrDateChange']
        }] }); })();
    return ClrDateInput;
}(WrappedFormControl));
export { ClrDateInput };
