// @clr/angular/esm5/forms/datepicker/day.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
import { ClrPopoverToggleService } from '../../utils/popover/providers/popover-toggle.service';
import { DateFormControlService } from './providers/date-form-control.service';
import { DateNavigationService } from './providers/date-navigation.service';
import * as ɵngcc0 from '@angular/core';
var ClrDay = /** @class */ (function () {
    function ClrDay(_dateNavigationService, _toggleService, dateFormControlService) {
        this._dateNavigationService = _dateNavigationService;
        this._toggleService = _toggleService;
        this.dateFormControlService = dateFormControlService;
    }
    Object.defineProperty(ClrDay.prototype, "dayView", {
        get: function () {
            return this._dayView;
        },
        /**
         * DayViewModel input which is used to build the Day View.
         */
        set: function (day) {
            this._dayView = day;
            this.dayString = this._dayView.dayModel.toDateString();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Updates the focusedDay in the DateNavigationService when the ClrDay is focused.
     */
    ClrDay.prototype.onDayViewFocus = function () {
        this._dateNavigationService.focusedDay = this.dayView.dayModel;
    };
    /**
     * Updates the selectedDay when the ClrDay is selected and closes the datepicker popover.
     */
    ClrDay.prototype.selectDay = function () {
        var day = this.dayView.dayModel;
        this._dateNavigationService.notifySelectedDayChanged(day);
        this.dateFormControlService.markAsDirty();
        this._toggleService.open = false;
    };
    ClrDay.ctorParameters = function () { return [
        { type: DateNavigationService },
        { type: ClrPopoverToggleService },
        { type: DateFormControlService }
    ]; };
    __decorate([
        Input('clrDayView')
    ], ClrDay.prototype, "dayView", null);
ClrDay.ɵfac = function ClrDay_Factory(t) { return new (t || ClrDay)(ɵngcc0.ɵɵdirectiveInject(DateNavigationService), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(DateFormControlService)); };
ClrDay.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrDay, selectors: [["clr-day"]], hostVars: 2, hostBindings: function ClrDay_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("day", true);
    } }, inputs: { dayView: ["clrDayView", "dayView"] }, decls: 2, vars: 11, consts: [["type", "button", 1, "day-btn", 3, "click", "focus"]], template: function ClrDay_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function ClrDay_Template_button_click_0_listener() { return ctx.selectDay(); })("focus", function ClrDay_Template_button_focus_0_listener() { return ctx.onDayViewFocus(); });
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("is-today", ctx.dayView.isTodaysDate)("is-excluded", ctx.dayView.isExcluded)("is-disabled", ctx.dayView.isDisabled)("is-selected", ctx.dayView.isSelected);
        ɵngcc0.ɵɵattribute("tabindex", ctx.dayView.tabIndex)("aria-label", ctx.dayString);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.dayView.dayModel.date, " ");
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrDay, [{
        type: Component,
        args: [{
                selector: 'clr-day',
                template: "\n        <button\n            class=\"day-btn\"\n            type=\"button\"\n            [class.is-today]=\"dayView.isTodaysDate\"\n            [class.is-excluded]=\"dayView.isExcluded\"\n            [class.is-disabled]=\"dayView.isDisabled\"\n            [class.is-selected]=\"dayView.isSelected\"\n            [attr.tabindex]=\"dayView.tabIndex\"\n            (click)=\"selectDay()\"\n            (focus)=\"onDayViewFocus()\"\n            [attr.aria-label]=\"dayString\">\n            {{dayView.dayModel.date}}\n        </button>\n    ",
                host: { '[class.day]': 'true' }
            }]
    }], function () { return [{ type: DateNavigationService }, { type: ClrPopoverToggleService }, { type: DateFormControlService }]; }, { dayView: [{
            type: Input,
            args: ['clrDayView']
        }] }); })();
    return ClrDay;
}());
export { ClrDay };
