// @clr/angular/esm5/popover/dropdown/providers/dropdown.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Optional, SkipSelf } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
var RootDropdownService = /** @class */ (function () {
    function RootDropdownService() {
        this._changes = new Subject();
    }
    Object.defineProperty(RootDropdownService.prototype, "changes", {
        get: function () {
            return this._changes.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    RootDropdownService.prototype.closeMenus = function () {
        this._changes.next(false);
    };
RootDropdownService.ɵfac = function RootDropdownService_Factory(t) { return new (t || RootDropdownService)(); };
RootDropdownService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: RootDropdownService, factory: function (t) { return RootDropdownService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RootDropdownService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return RootDropdownService;
}());
export { RootDropdownService };
export function clrRootDropdownFactory(existing) {
    return existing || new RootDropdownService();
}
export var ROOT_DROPDOWN_PROVIDER = {
    provide: RootDropdownService,
    useFactory: clrRootDropdownFactory,
    deps: [[new Optional(), new SkipSelf(), RootDropdownService]],
};
