// @clr/angular/esm5/utils/drag-and-drop/providers/draggable-snapshot.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { DomAdapter } from '../../dom-adapter/dom-adapter';
// This service is used to capture the state of clrDraggable element
// at a certain event and passes it to clrDraggableGhost component.
import * as ɵngcc0 from '@angular/core';
var DraggableSnapshotService = /** @class */ (function () {
    function DraggableSnapshotService(domAdapter) {
        this.domAdapter = domAdapter;
    }
    DraggableSnapshotService.prototype.capture = function (el, event) {
        this.draggableElClientRect = this.domAdapter.clientRect(el);
        this.snapshotDragEvent = event;
    };
    DraggableSnapshotService.prototype.discard = function () {
        delete this.draggableElClientRect;
        delete this.snapshotDragEvent;
    };
    Object.defineProperty(DraggableSnapshotService.prototype, "hasDraggableState", {
        get: function () {
            return !!this.snapshotDragEvent && !!this.draggableElClientRect;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DraggableSnapshotService.prototype, "clientRect", {
        get: function () {
            return this.draggableElClientRect;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DraggableSnapshotService.prototype, "dragEvent", {
        get: function () {
            return this.snapshotDragEvent;
        },
        enumerable: true,
        configurable: true
    });
    DraggableSnapshotService.ctorParameters = function () { return [
        { type: DomAdapter }
    ]; };
DraggableSnapshotService.ɵfac = function DraggableSnapshotService_Factory(t) { return new (t || DraggableSnapshotService)(ɵngcc0.ɵɵinject(DomAdapter)); };
DraggableSnapshotService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DraggableSnapshotService, factory: function (t) { return DraggableSnapshotService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DraggableSnapshotService, [{
        type: Injectable
    }], function () { return [{ type: DomAdapter }]; }, null); })();
    return DraggableSnapshotService;
}());
export { DraggableSnapshotService };
