// @clr/angular/esm5/utils/drag-and-drop/providers/global-drag-mode.service.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Renderer2 } from '@angular/core';
// This service class adds and removes the "in-drag" class to the document body element
// through its public enter() and exit() methods.
import * as ɵngcc0 from '@angular/core';
var GlobalDragModeService = /** @class */ (function () {
    function GlobalDragModeService(renderer) {
        this.renderer = renderer;
    }
    GlobalDragModeService.prototype.enter = function () {
        this.renderer.addClass(document.body, 'in-drag');
    };
    GlobalDragModeService.prototype.exit = function () {
        this.renderer.removeClass(document.body, 'in-drag');
    };
    GlobalDragModeService.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
GlobalDragModeService.ɵfac = function GlobalDragModeService_Factory(t) { return new (t || GlobalDragModeService)(ɵngcc0.ɵɵinject(ɵngcc0.Renderer2)); };
GlobalDragModeService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GlobalDragModeService, factory: function (t) { return GlobalDragModeService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GlobalDragModeService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, null); })();
    return GlobalDragModeService;
}());
export { GlobalDragModeService };
