// @clr/angular/esm5/utils/focus/focus-on-view-init/focus-on-view-init.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { DOCUMENT, isPlatformBrowser } from '@angular/common';
import { AfterViewInit, Directive, ElementRef, HostListener, Inject, Input, PLATFORM_ID, Renderer2, } from '@angular/core';
import { FOCUS_ON_VIEW_INIT } from './focus-on-view-init.provider';
/*  This directive is for guiding the document focus to the newly added content when its view is initialized
    so that assistive technologies can read its content. */
import * as ɵngcc0 from '@angular/core';
var ClrFocusOnViewInit = /** @class */ (function () {
    function ClrFocusOnViewInit(el, platformId, focusOnViewInit, document, renderer) {
        this.el = el;
        this.platformId = platformId;
        this.focusOnViewInit = focusOnViewInit;
        this.renderer = renderer;
        this.directFocus = true; // true if the element gets focused without need to set tabindex;
        this._isEnabled = this.focusOnViewInit;
        // Angular compiler doesn't understand the type Document
        // when working out the metadata for injectable parameters,
        // even though it understands the injection token DOCUMENT
        // https://github.com/angular/angular/issues/20351
        this.document = document;
    }
    ClrFocusOnViewInit.prototype.onFocusout = function () {
        if (!this.directFocus) {
            // manually set attributes and styles should be removed
            this.renderer.removeAttribute(this.el.nativeElement, 'tabindex');
            this.renderer.setStyle(this.el.nativeElement, 'outline', null);
        }
    };
    Object.defineProperty(ClrFocusOnViewInit.prototype, "isEnabled", {
        set: function (value) {
            if (this.focusOnViewInit && typeof value === 'boolean') {
                this._isEnabled = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    ClrFocusOnViewInit.prototype.ngAfterViewInit = function () {
        this.focus();
    };
    ClrFocusOnViewInit.prototype.focus = function () {
        if (!isPlatformBrowser(this.platformId)) {
            return;
        }
        if (!this._isEnabled) {
            return;
        }
        if (this.document && this.document.activeElement !== this.el.nativeElement) {
            this.el.nativeElement.focus();
            if (this.document.activeElement !== this.el.nativeElement) {
                // if it's not directly focused now, it means it was a non-interactive element
                // so we need to give it a tabindex.
                this.directFocus = false;
                this.renderer.setAttribute(this.el.nativeElement, 'tabindex', '-1');
                this.renderer.setStyle(this.el.nativeElement, 'outline', 'none');
                this.el.nativeElement.focus();
            }
        }
    };
    ClrFocusOnViewInit.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [FOCUS_ON_VIEW_INIT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: Renderer2 }
    ]; };
    __decorate([
        HostListener('focusout')
    ], ClrFocusOnViewInit.prototype, "onFocusout", null);
    __decorate([
        Input('clrFocusOnViewInit')
    ], ClrFocusOnViewInit.prototype, "isEnabled", null);
    ClrFocusOnViewInit = __decorate([ __param(1, Inject(PLATFORM_ID)),
        __param(2, Inject(FOCUS_ON_VIEW_INIT)),
        __param(3, Inject(DOCUMENT))
    ], ClrFocusOnViewInit);
ClrFocusOnViewInit.ɵfac = function ClrFocusOnViewInit_Factory(t) { return new (t || ClrFocusOnViewInit)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID), ɵngcc0.ɵɵdirectiveInject(FOCUS_ON_VIEW_INIT), ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
ClrFocusOnViewInit.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrFocusOnViewInit, selectors: [["", "clrFocusOnViewInit", ""]], hostBindings: function ClrFocusOnViewInit_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("focusout", function ClrFocusOnViewInit_focusout_HostBindingHandler() { return ctx.onFocusout(); });
    } }, inputs: { isEnabled: ["clrFocusOnViewInit", "isEnabled"] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrFocusOnViewInit, [{
        type: Directive,
        args: [{
                selector: '[clrFocusOnViewInit]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }, { type: Boolean, decorators: [{
                type: Inject,
                args: [FOCUS_ON_VIEW_INIT]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc0.Renderer2 }]; }, { onFocusout: [{
            type: HostListener,
            args: ['focusout']
        }], isEnabled: [{
            type: Input,
            args: ['clrFocusOnViewInit']
        }] }); })();
    return ClrFocusOnViewInit;
}());
export { ClrFocusOnViewInit };
