// @clr/angular/esm5/utils/focus/focus-on-view-init/focus-on-view-init.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrFocusOnViewInit } from './focus-on-view-init';
import { FOCUS_ON_VIEW_INIT_PROVIDER } from './focus-on-view-init.provider';
import * as ɵngcc0 from '@angular/core';
export var FOCUS_ON_VIEW_INIT_DIRECTIVES = [ClrFocusOnViewInit];
var ClrFocusOnViewInitModule = /** @class */ (function () {
    function ClrFocusOnViewInitModule() {
    }
ClrFocusOnViewInitModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrFocusOnViewInitModule });
ClrFocusOnViewInitModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrFocusOnViewInitModule_Factory(t) { return new (t || ClrFocusOnViewInitModule)(); }, providers: [FOCUS_ON_VIEW_INIT_PROVIDER], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrFocusOnViewInitModule, { declarations: function () { return [ClrFocusOnViewInit]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrFocusOnViewInit]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrFocusOnViewInitModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [FOCUS_ON_VIEW_INIT_DIRECTIVES],
                providers: [FOCUS_ON_VIEW_INIT_PROVIDER],
                exports: [FOCUS_ON_VIEW_INIT_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrFocusOnViewInitModule;
}());
export { ClrFocusOnViewInitModule };
