// @clr/angular/esm5/forms/checkbox/checkbox.module.js
/**
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrHostWrappingModule } from '../../utils/host-wrapping/host-wrapping.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrCheckbox } from './checkbox';
import { ClrCheckboxContainer } from './checkbox-container';
import { ClrCheckboxWrapper } from './checkbox-wrapper';
import * as ɵngcc0 from '@angular/core';
var ClrCheckboxModule = /** @class */ (function () {
    function ClrCheckboxModule() {
    }
ClrCheckboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrCheckboxModule });
ClrCheckboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrCheckboxModule_Factory(t) { return new (t || ClrCheckboxModule)(); }, imports: [[CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrCheckboxModule, { declarations: function () { return [ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper]; }, imports: function () { return [CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule]; }, exports: function () { return [ClrCommonFormsModule, ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrCheckboxModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrCommonFormsModule, ClrHostWrappingModule],
                declarations: [ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper],
                exports: [ClrCommonFormsModule, ClrCheckbox, ClrCheckboxContainer, ClrCheckboxWrapper],
                entryComponents: [ClrCheckboxWrapper]
            }]
    }], function () { return []; }, null); })();
    return ClrCheckboxModule;
}());
export { ClrCheckboxModule };
