// @clr/angular/esm2015/utils/scrolling/scrolling-service.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let ScrollingService = class ScrollingService {
    constructor(_document) {
        this._document = _document;
    }
    stopScrolling() {
        this._document.body.classList.add('no-scrolling');
    }
    resumeScrolling() {
        if (this._document.body.classList.contains('no-scrolling')) {
            this._document.body.classList.remove('no-scrolling');
        }
    }
};
ScrollingService.ɵfac = function ScrollingService_Factory(t) { return new (t || ScrollingService)(ɵngcc0.ɵɵinject(DOCUMENT)); };
ScrollingService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ScrollingService, factory: ScrollingService.ɵfac });
ScrollingService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
ScrollingService = __decorate([ __param(0, Inject(DOCUMENT))
], ScrollingService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ScrollingService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, null); })();
export { ScrollingService };
