// @clr/angular/esm2015/accordion/accordion-title.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
let ClrAccordionTitle = class ClrAccordionTitle {
};
ClrAccordionTitle.ɵfac = function ClrAccordionTitle_Factory(t) { return new (t || ClrAccordionTitle)(); };
ClrAccordionTitle.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAccordionTitle, selectors: [["clr-accordion-title"], ["clr-step-title"]], hostVars: 2, hostBindings: function ClrAccordionTitle_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-accordion-title", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrAccordionTitle_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAccordionTitle, [{
        type: Component,
        args: [{
                selector: 'clr-accordion-title, clr-step-title',
                template: `<ng-content></ng-content>`,
                host: { '[class.clr-accordion-title]': 'true' },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
export { ClrAccordionTitle };
