// @clr/angular/esm2015/accordion/accordion-description.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
let ClrAccordionDescription = class ClrAccordionDescription {
};
ClrAccordionDescription.ɵfac = function ClrAccordionDescription_Factory(t) { return new (t || ClrAccordionDescription)(); };
ClrAccordionDescription.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAccordionDescription, selectors: [["clr-accordion-description"], ["clr-step-description"]], hostVars: 2, hostBindings: function ClrAccordionDescription_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-accordion-description", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrAccordionDescription_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAccordionDescription, [{
        type: Component,
        args: [{
                selector: 'clr-accordion-description, clr-step-description',
                template: `<ng-content></ng-content>`,
                host: { '[class.clr-accordion-description]': 'true' },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
export { ClrAccordionDescription };
