// @clr/angular/esm2015/utils/popover/popover-close-button.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { __decorate } from "tslib";
import { Directive, ElementRef, EventEmitter, Output, HostListener, OnDestroy, AfterViewInit } from '@angular/core';
import { ClrPopoverEventsService } from './providers/popover-events.service';
import { ClrPopoverToggleService } from './providers/popover-toggle.service';
import { filter } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
let ClrPopoverCloseButton = class ClrPopoverCloseButton {
    constructor(elementRef, smartEventsService, smartOpenService) {
        this.elementRef = elementRef;
        this.smartEventsService = smartEventsService;
        this.smartOpenService = smartOpenService;
        this.subscriptions = [];
        this.closeChange = new EventEmitter();
        this.subscriptions.push(smartOpenService.openChange.pipe(filter(value => !value)).subscribe(() => {
            this.closeChange.next();
        }));
    }
    handleClick(event) {
        this.smartOpenService.toggleWithEvent(event);
        this.smartEventsService.setAnchorFocus();
    }
    ngAfterViewInit() {
        this.smartEventsService.closeButtonRef = this.elementRef;
        this.smartEventsService.setCloseFocus();
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
};
ClrPopoverCloseButton.ɵfac = function ClrPopoverCloseButton_Factory(t) { return new (t || ClrPopoverCloseButton)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ClrPopoverEventsService), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService)); };
ClrPopoverCloseButton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrPopoverCloseButton, selectors: [["", "clrPopoverCloseButton", ""]], hostVars: 2, hostBindings: function ClrPopoverCloseButton_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrPopoverCloseButton_click_HostBindingHandler($event) { return ctx.handleClick($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-smart-close-button", true);
    } }, outputs: { closeChange: "clrPopoverOnCloseChange" } });
ClrPopoverCloseButton.ctorParameters = () => [
    { type: ElementRef },
    { type: ClrPopoverEventsService },
    { type: ClrPopoverToggleService }
];
__decorate([
    Output('clrPopoverOnCloseChange')
], ClrPopoverCloseButton.prototype, "closeChange", void 0);
__decorate([
    HostListener('click', ['$event'])
], ClrPopoverCloseButton.prototype, "handleClick", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrPopoverCloseButton, [{
        type: Directive,
        args: [{
                selector: '[clrPopoverCloseButton]',
                host: {
                    '[class.clr-smart-close-button]': 'true'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ClrPopoverEventsService }, { type: ClrPopoverToggleService }]; }, { closeChange: [{
            type: Output,
            args: ['clrPopoverOnCloseChange']
        }], handleClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
export { ClrPopoverCloseButton };
