// @clr/angular/esm2015/utils/popover/popover-open-close-button.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { __decorate } from "tslib";
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { ClrPopoverToggleService } from './providers/popover-toggle.service';
import * as ɵngcc0 from '@angular/core';
let ClrPopoverOpenCloseButton = class ClrPopoverOpenCloseButton {
    constructor(smartOpenService) {
        this.smartOpenService = smartOpenService;
        this.subscriptions = [];
        this.openCloseChange = new EventEmitter();
        this.subscriptions.push(this.smartOpenService.openChange.subscribe(change => {
            this.openCloseChange.next(change);
        }));
    }
    handleClick(event) {
        this.smartOpenService.toggleWithEvent(event);
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
};
ClrPopoverOpenCloseButton.ɵfac = function ClrPopoverOpenCloseButton_Factory(t) { return new (t || ClrPopoverOpenCloseButton)(ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService)); };
ClrPopoverOpenCloseButton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrPopoverOpenCloseButton, selectors: [["", "clrPopoverOpenCloseButton", ""]], hostVars: 2, hostBindings: function ClrPopoverOpenCloseButton_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClrPopoverOpenCloseButton_click_HostBindingHandler($event) { return ctx.handleClick($event); });
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-smart-open-close", true);
    } }, outputs: { openCloseChange: "clrPopoverOpenCloseChange" } });
ClrPopoverOpenCloseButton.ctorParameters = () => [
    { type: ClrPopoverToggleService }
];
__decorate([
    Output('clrPopoverOpenCloseChange')
], ClrPopoverOpenCloseButton.prototype, "openCloseChange", void 0);
__decorate([
    HostListener('click', ['$event'])
], ClrPopoverOpenCloseButton.prototype, "handleClick", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrPopoverOpenCloseButton, [{
        type: Directive,
        args: [{
                selector: '[clrPopoverOpenCloseButton]',
                host: {
                    '[class.clr-smart-open-close]': 'true'
                }
            }]
    }], function () { return [{ type: ClrPopoverToggleService }]; }, { openCloseChange: [{
            type: Output,
            args: ['clrPopoverOpenCloseChange']
        }], handleClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
export { ClrPopoverOpenCloseButton };
