// @clr/angular/esm5/layout/tabs/tab.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ContentChild, Inject } from '@angular/core';
import { IF_ACTIVE_ID, IF_ACTIVE_ID_PROVIDER, IfActiveService } from '../../utils/conditional/if-active.service';
import { AriaService } from './providers/aria.service';
import { TabsService } from './providers/tabs.service';
import { ClrTabContent } from './tab-content';
import { ClrTabLink } from './tab-link.directive';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrTab = /** @class */ (function () {
    function ClrTab(ifActiveService, id, tabsService) {
        this.ifActiveService = ifActiveService;
        this.id = id;
        this.tabsService = tabsService;
        tabsService.register(this);
    }
    ClrTab.prototype.ngOnDestroy = function () {
        this.tabsService.unregister(this);
    };
    Object.defineProperty(ClrTab.prototype, "active", {
        get: function () {
            return this.ifActiveService.current === this.id;
        },
        enumerable: true,
        configurable: true
    });
    ClrTab.ctorParameters = function () { return [
        { type: IfActiveService },
        { type: Number, decorators: [{ type: Inject, args: [IF_ACTIVE_ID,] }] },
        { type: TabsService }
    ]; };
    __decorate([
        ContentChild(ClrTabLink, { static: true })
    ], ClrTab.prototype, "tabLink", void 0);
    __decorate([
        ContentChild(ClrTabContent, { static: true })
    ], ClrTab.prototype, "tabContent", void 0);
    ClrTab = __decorate([ __param(1, Inject(IF_ACTIVE_ID))
    ], ClrTab);
ClrTab.ɵfac = function ClrTab_Factory(t) { return new (t || ClrTab)(ɵngcc0.ɵɵdirectiveInject(IfActiveService), ɵngcc0.ɵɵdirectiveInject(IF_ACTIVE_ID), ɵngcc0.ɵɵdirectiveInject(TabsService)); };
ClrTab.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTab, selectors: [["clr-tab"]], contentQueries: function ClrTab_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, ClrTabLink, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, ClrTabContent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabLink = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabContent = _t.first);
    } }, features: [ɵngcc0.ɵɵProvidersFeature([IF_ACTIVE_ID_PROVIDER, AriaService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTab_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTab, [{
        type: Component,
        args: [{
                selector: 'clr-tab',
                template: "\n        <ng-content></ng-content>\n    ",
                providers: [IF_ACTIVE_ID_PROVIDER, AriaService]
            }]
    }], function () { return [{ type: IfActiveService }, { type: Number, decorators: [{
                type: Inject,
                args: [IF_ACTIVE_ID]
            }] }, { type: TabsService }]; }, { tabLink: [{
            type: ContentChild,
            args: [ClrTabLink, { static: true }]
        }], tabContent: [{
            type: ContentChild,
            args: [ClrTabContent, { static: true }]
        }] }); })();
    return ClrTab;
}());
export { ClrTab };
