// @clr/angular/esm2015/utils/loading/loading.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, Input, Optional } from '@angular/core';
import { LoadingListener } from './loading-listener';
import * as ɵngcc0 from '@angular/core';
export var ClrLoadingState;
(function (ClrLoadingState) {
    ClrLoadingState[ClrLoadingState["DEFAULT"] = 0] = "DEFAULT";
    ClrLoadingState[ClrLoadingState["LOADING"] = 1] = "LOADING";
    ClrLoadingState[ClrLoadingState["SUCCESS"] = 2] = "SUCCESS";
    ClrLoadingState[ClrLoadingState["ERROR"] = 3] = "ERROR";
})(ClrLoadingState || (ClrLoadingState = {}));
let ClrLoading = class ClrLoading {
    // We find the first parent that handles something loading
    constructor(listener) {
        this.listener = listener;
        this._loadingState = ClrLoadingState.DEFAULT;
    }
    get loadingState() {
        return this._loadingState;
    }
    set loadingState(value) {
        if (value === true) {
            value = ClrLoadingState.LOADING;
        }
        else if (!value) {
            value = ClrLoadingState.DEFAULT;
        }
        if (value === this._loadingState) {
            return;
        }
        this._loadingState = value;
        if (this.listener) {
            this.listener.loadingStateChange(value);
        }
    }
    ngOnDestroy() {
        this.loadingState = ClrLoadingState.DEFAULT;
    }
};
ClrLoading.ɵfac = function ClrLoading_Factory(t) { return new (t || ClrLoading)(ɵngcc0.ɵɵdirectiveInject(LoadingListener, 8)); };
ClrLoading.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrLoading, selectors: [["", "clrLoading", ""]], inputs: { loadingState: ["clrLoading", "loadingState"] } });
ClrLoading.ctorParameters = () => [
    { type: LoadingListener, decorators: [{ type: Optional }] }
];
__decorate([
    Input('clrLoading')
], ClrLoading.prototype, "loadingState", null);
ClrLoading = __decorate([ __param(0, Optional())
], ClrLoading);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrLoading, [{
        type: Directive,
        args: [{ selector: '[clrLoading]' }]
    }], function () { return [{ type: LoadingListener, decorators: [{
                type: Optional
            }] }]; }, { loadingState: [{
            type: Input,
            args: ['clrLoading']
        }] }); })();
export { ClrLoading };
