// @clr/angular/esm5/layout/vertical-nav/vertical-nav.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrFocusOnViewInitModule } from '../../utils/focus/focus-on-view-init/focus-on-view-init.module';
import { ClrVerticalNav } from './vertical-nav';
import { ClrVerticalNavGroup } from './vertical-nav-group';
import { ClrVerticalNavGroupChildren } from './vertical-nav-group-children';
import { ClrVerticalNavIcon } from './vertical-nav-icon';
import { ClrVerticalNavLink } from './vertical-nav-link';
import * as ɵngcc0 from '@angular/core';
export var CLR_VERTICAL_NAV_DIRECTIVES = [
    ClrVerticalNav,
    ClrVerticalNavLink,
    ClrVerticalNavGroup,
    ClrVerticalNavGroupChildren,
    ClrVerticalNavIcon,
];
var ClrVerticalNavModule = /** @class */ (function () {
    function ClrVerticalNavModule() {
    }
ClrVerticalNavModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrVerticalNavModule });
ClrVerticalNavModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrVerticalNavModule_Factory(t) { return new (t || ClrVerticalNavModule)(); }, imports: [[CommonModule, ClrIconModule, ClrConditionalModule, ClrFocusOnViewInitModule],
        ClrConditionalModule, ClrIconModule, ClrFocusOnViewInitModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrVerticalNavModule, { declarations: function () { return [ClrVerticalNav,
        ClrVerticalNavLink,
        ClrVerticalNavGroup,
        ClrVerticalNavGroupChildren,
        ClrVerticalNavIcon]; }, imports: function () { return [CommonModule, ClrIconModule, ClrConditionalModule, ClrFocusOnViewInitModule]; }, exports: function () { return [ClrVerticalNav,
        ClrVerticalNavLink,
        ClrVerticalNavGroup,
        ClrVerticalNavGroupChildren,
        ClrVerticalNavIcon,
        ClrConditionalModule, ClrIconModule, ClrFocusOnViewInitModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrVerticalNavModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrConditionalModule, ClrFocusOnViewInitModule],
                declarations: [CLR_VERTICAL_NAV_DIRECTIVES],
                exports: [CLR_VERTICAL_NAV_DIRECTIVES, ClrConditionalModule, ClrIconModule, ClrFocusOnViewInitModule]
            }]
    }], function () { return []; }, null); })();
    return ClrVerticalNavModule;
}());
export { ClrVerticalNavModule };
