// @clr/angular/esm5/layout/vertical-nav/vertical-nav-group-children.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrVerticalNavGroupChildren = /** @class */ (function () {
    function ClrVerticalNavGroupChildren() {
    }
ClrVerticalNavGroupChildren.ɵfac = function ClrVerticalNavGroupChildren_Factory(t) { return new (t || ClrVerticalNavGroupChildren)(); };
ClrVerticalNavGroupChildren.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrVerticalNavGroupChildren, selectors: [["clr-vertical-nav-group-children"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrVerticalNavGroupChildren_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrVerticalNavGroupChildren, [{
        type: Component,
        args: [{
                selector: 'clr-vertical-nav-group-children',
                template: "\n        <ng-content></ng-content>\n    "
            }]
    }], function () { return []; }, null); })();
    return ClrVerticalNavGroupChildren;
}());
export { ClrVerticalNavGroupChildren };
