// @clr/angular/esm5/wizard/wizard-header-action.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var wizardHeaderActionIndex = 0;
var ClrWizardHeaderAction = /** @class */ (function () {
    function ClrWizardHeaderAction() {
        // title is explanatory text added to the header action
        this.title = '';
        // If our host has an ID attribute, we use this instead of our index.
        this._id = (wizardHeaderActionIndex++).toString();
        this.disabled = false;
        this.headerActionClicked = new EventEmitter(false);
    }
    Object.defineProperty(ClrWizardHeaderAction.prototype, "id", {
        get: function () {
            return "clr-wizard-header-action-" + this._id;
        },
        enumerable: true,
        configurable: true
    });
    ClrWizardHeaderAction.prototype.click = function () {
        if (this.disabled) {
            return;
        }
        // passing the header action id allows users to have one method that
        // routes to many different actions based on the type of header action
        // clicked. this is further aided by users being able to specify ids
        // for their header actions.
        this.headerActionClicked.emit(this._id);
    };
    __decorate([
        Input('title')
    ], ClrWizardHeaderAction.prototype, "title", void 0);
    __decorate([
        Input('id')
    ], ClrWizardHeaderAction.prototype, "_id", void 0);
    __decorate([
        Input('clrWizardHeaderActionDisabled')
    ], ClrWizardHeaderAction.prototype, "disabled", void 0);
    __decorate([
        Output('actionClicked')
    ], ClrWizardHeaderAction.prototype, "headerActionClicked", void 0);
ClrWizardHeaderAction.ɵfac = function ClrWizardHeaderAction_Factory(t) { return new (t || ClrWizardHeaderAction)(); };
ClrWizardHeaderAction.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrWizardHeaderAction, selectors: [["clr-wizard-header-action"]], hostAttrs: [1, "clr-wizard-header-action-wrapper"], inputs: { title: "title", _id: ["id", "_id"], disabled: ["clrWizardHeaderActionDisabled", "disabled"] }, outputs: { headerActionClicked: "actionClicked" }, ngContentSelectors: _c0, decls: 2, vars: 4, consts: [["type", "button", 1, "btn", "clr-wizard-header-action", "btn-link", 3, "id", "title", "click"]], template: function ClrWizardHeaderAction_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function ClrWizardHeaderAction_Template_button_click_0_listener() { return ctx.click(); });
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("disabled", ctx.disabled);
        ɵngcc0.ɵɵproperty("id", ctx.id)("title", ctx.title);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrWizardHeaderAction, [{
        type: Component,
        args: [{
                selector: 'clr-wizard-header-action',
                template: "\n        <button \n            type=\"button\"\n            class=\"btn clr-wizard-header-action btn-link\"\n            [id]=\"id\"\n            [class.disabled]=\"disabled\"\n            (click)=\"click()\"\n            [title]=\"title\">\n            <ng-content></ng-content>\n        </button>\n    ",
                host: { class: 'clr-wizard-header-action-wrapper' }
            }]
    }], function () { return []; }, { title: [{
            type: Input,
            args: ['title']
        }], _id: [{
            type: Input,
            args: ['id']
        }], disabled: [{
            type: Input,
            args: ['clrWizardHeaderActionDisabled']
        }], headerActionClicked: [{
            type: Output,
            args: ['actionClicked']
        }] }); })();
    return ClrWizardHeaderAction;
}());
export { ClrWizardHeaderAction };
