// @clr/angular/esm5/wizard/wizard-stepnav.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { PageCollectionService } from './providers/page-collection.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from './wizard-stepnav-item';

function ClrWizardStepnav_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var page_r1 = ctx.$implicit;
    var i_r2 = ctx.index;
    ɵngcc0.ɵɵproperty("page", page_r1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(i_r2 + 1);
} }
var ClrWizardStepnav = /** @class */ (function () {
    function ClrWizardStepnav(pageService) {
        this.pageService = pageService;
    }
    ClrWizardStepnav.ctorParameters = function () { return [
        { type: PageCollectionService }
    ]; };
ClrWizardStepnav.ɵfac = function ClrWizardStepnav_Factory(t) { return new (t || ClrWizardStepnav)(ɵngcc0.ɵɵdirectiveInject(PageCollectionService)); };
ClrWizardStepnav.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrWizardStepnav, selectors: [["clr-wizard-stepnav"]], hostAttrs: [1, "clr-wizard-stepnav"], decls: 2, vars: 1, consts: [[1, "clr-wizard-stepnav-list"], ["clr-wizard-stepnav-item", "", "class", "clr-wizard-stepnav-item", 3, "page", 4, "ngFor", "ngForOf"], ["clr-wizard-stepnav-item", "", 1, "clr-wizard-stepnav-item", 3, "page"]], template: function ClrWizardStepnav_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ClrWizardStepnav_div_1_Template, 2, 2, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.pageService.pages);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc2.ClrWizardStepnavItem], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrWizardStepnav, [{
        type: Component,
        args: [{
                selector: 'clr-wizard-stepnav',
                template: "\n    <div class=\"clr-wizard-stepnav-list\">\n      <div *ngFor=\"let page of pageService.pages; let i = index\" clr-wizard-stepnav-item [page]=\"page\" class=\"clr-wizard-stepnav-item\">{{i + 1}}</div>\n    </div>\n  ",
                host: { class: 'clr-wizard-stepnav' }
            }]
    }], function () { return [{ type: PageCollectionService }]; }, null); })();
    return ClrWizardStepnav;
}());
export { ClrWizardStepnav };
