// @clr/angular/esm5/forms/common/providers/control-class.service.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Injectable, Optional } from '@angular/core';
import { LayoutService } from './layout.service';
import * as ɵngcc0 from '@angular/core';
var ControlClassService = /** @class */ (function () {
    function ControlClassService(layoutService) {
        this.layoutService = layoutService;
        this.className = '';
    }
    ControlClassService.prototype.controlClass = function (invalid, grid, additional) {
        if (invalid === void 0) { invalid = false; }
        if (grid === void 0) { grid = false; }
        if (additional === void 0) { additional = ''; }
        var controlClasses = [this.className, additional];
        if (invalid) {
            controlClasses.push('clr-error');
        }
        if (grid && this.layoutService && this.className.indexOf('clr-col') === -1) {
            controlClasses.push("clr-col-md-" + (this.layoutService.maxLabelSize - this.layoutService.labelSize) + " clr-col-12");
        }
        return controlClasses.join(' ').trim();
    };
    // We want to remove the column classes from the input up to the container
    ControlClassService.prototype.initControlClass = function (renderer, element) {
        if (element && element.className) {
            this.className = element.className;
            var klasses = element.className.split(' ');
            klasses.forEach(function (klass) {
                if (klass.startsWith('clr-col')) {
                    renderer.removeClass(element, klass);
                }
            });
        }
    };
    ControlClassService.ctorParameters = function () { return [
        { type: LayoutService, decorators: [{ type: Optional }] }
    ]; };
    ControlClassService = __decorate([ __param(0, Optional())
    ], ControlClassService);
ControlClassService.ɵfac = function ControlClassService_Factory(t) { return new (t || ControlClassService)(ɵngcc0.ɵɵinject(LayoutService, 8)); };
ControlClassService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ControlClassService, factory: function (t) { return ControlClassService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ControlClassService, [{
        type: Injectable
    }], function () { return [{ type: LayoutService, decorators: [{
                type: Optional
            }] }]; }, null); })();
    return ControlClassService;
}());
export { ControlClassService };
