// @clr/angular/esm5/forms/common/abstract-container.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { ContentChild, Directive, Optional } from '@angular/core';
import { IfErrorService } from './if-error/if-error.service';
import { NgControlService } from './providers/ng-control.service';
import { LayoutService } from './providers/layout.service';
import { ClrLabel } from './label';
import { ControlClassService } from './providers/control-class.service';
import * as ɵngcc0 from '@angular/core';
var ClrAbstractContainer = /** @class */ (function () {
    function ClrAbstractContainer(ifErrorService, layoutService, controlClassService, ngControlService) {
        var _this = this;
        this.ifErrorService = ifErrorService;
        this.layoutService = layoutService;
        this.controlClassService = controlClassService;
        this.ngControlService = ngControlService;
        this.subscriptions = [];
        this.invalid = false;
        this._dynamic = false;
        this.subscriptions.push(this.ifErrorService.statusChanges.subscribe(function (invalid) {
            _this.invalid = invalid;
        }));
        this.subscriptions.push(this.ngControlService.controlChanges.subscribe(function (control) {
            _this.control = control;
        }));
    }
    ClrAbstractContainer.prototype.controlClass = function () {
        return this.controlClassService.controlClass(this.invalid, this.addGrid());
    };
    ClrAbstractContainer.prototype.addGrid = function () {
        return this.layoutService && !this.layoutService.isVertical();
    };
    ClrAbstractContainer.prototype.ngOnDestroy = function () {
        this.subscriptions.map(function (sub) { return sub.unsubscribe(); });
    };
    ClrAbstractContainer.ctorParameters = function () { return [
        { type: IfErrorService },
        { type: LayoutService, decorators: [{ type: Optional }] },
        { type: ControlClassService },
        { type: NgControlService }
    ]; };
    __decorate([
        ContentChild(ClrLabel, { static: false })
    ], ClrAbstractContainer.prototype, "label", void 0);
    ClrAbstractContainer = __decorate([ __param(1, Optional())
    ], ClrAbstractContainer);
ClrAbstractContainer.ɵfac = function ClrAbstractContainer_Factory(t) { return new (t || ClrAbstractContainer)(ɵngcc0.ɵɵdirectiveInject(IfErrorService), ɵngcc0.ɵɵdirectiveInject(LayoutService, 8), ɵngcc0.ɵɵdirectiveInject(ControlClassService), ɵngcc0.ɵɵdirectiveInject(NgControlService)); };
ClrAbstractContainer.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrAbstractContainer, contentQueries: function ClrAbstractContainer_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrLabel, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.label = _t.first);
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAbstractContainer, [{
        type: Directive
    }], function () { return [{ type: IfErrorService }, { type: LayoutService, decorators: [{
                type: Optional
            }] }, { type: ControlClassService }, { type: NgControlService }]; }, { label: [{
            type: ContentChild,
            args: [ClrLabel, { static: false }]
        }] }); })();
    return ClrAbstractContainer;
}());
export { ClrAbstractContainer };
