// @clr/angular/esm5/forms/common/common.module.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrControlError } from './error';
import { ClrControlHelper } from './helper';
import { ClrIfError } from './if-error/if-error';
import { ClrLabel } from './label';
import { ClrForm } from './form';
import { ClrLayout } from './layout';
import { ClrControlContainer } from './control-container';
import { ClrControl } from './control';
import * as ɵngcc0 from '@angular/core';
var ClrCommonFormsModule = /** @class */ (function () {
    function ClrCommonFormsModule() {
    }
ClrCommonFormsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrCommonFormsModule });
ClrCommonFormsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrCommonFormsModule_Factory(t) { return new (t || ClrCommonFormsModule)(); }, imports: [[CommonModule, ClrIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrCommonFormsModule, { declarations: function () { return [ClrLabel,
        ClrControlError,
        ClrControlHelper,
        ClrIfError,
        ClrForm,
        ClrLayout,
        ClrControlContainer,
        ClrControl]; }, imports: function () { return [CommonModule, ClrIconModule]; }, exports: function () { return [ClrLabel,
        ClrControlError,
        ClrControlHelper,
        ClrIfError,
        ClrForm,
        ClrLayout,
        ClrControlContainer,
        ClrControl]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrCommonFormsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule],
                declarations: [
                    ClrLabel,
                    ClrControlError,
                    ClrControlHelper,
                    ClrIfError,
                    ClrForm,
                    ClrLayout,
                    ClrControlContainer,
                    ClrControl,
                ],
                exports: [
                    ClrLabel,
                    ClrControlError,
                    ClrControlHelper,
                    ClrIfError,
                    ClrForm,
                    ClrLayout,
                    ClrControlContainer,
                    ClrControl,
                ],
                entryComponents: [ClrControlContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrCommonFormsModule;
}());
export { ClrCommonFormsModule };
