// @clr/angular/esm5/forms/common/error.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Component, Optional, ElementRef, AfterViewInit } from '@angular/core';
import { ControlIdService } from './providers/control-id.service';
import { ClrAriaLiveService } from '../../utils/a11y/aria-live.service';
import { ControlClassService } from './providers/control-class.service';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrControlError = /** @class */ (function () {
    function ClrControlError(controlIdService, controlClassService, ariaLiveService, el) {
        this.controlIdService = controlIdService;
        this.controlClassService = controlClassService;
        this.ariaLiveService = ariaLiveService;
        this.el = el;
    }
    /** @deprecated since 3.0, remove in 4.0 - ariaLiveService */
    ClrControlError.prototype.ngAfterViewInit = function () {
        /**
         * The way we render elements inside the `clr-control-container make this
         * component announce itself without been visible on the screen.
         *
         * The check below try to guess is clr-controll-error used in some of the
         * cases mention above and prevent us from announcing without the need of that.
         *
         * This change won't create breaking change - but will make aria live announcment
         * work only when the component is used outside forms or some of the components.
         *
         * This is temporary solution - until better is found.
         *
         */
        if (this.controlClassService === null) {
            this.ariaLiveService.announce(this.el.nativeElement);
        }
    };
    ClrControlError.ctorParameters = function () { return [
        { type: ControlIdService, decorators: [{ type: Optional }] },
        { type: ControlClassService, decorators: [{ type: Optional }] },
        { type: ClrAriaLiveService },
        { type: ElementRef }
    ]; };
    ClrControlError = __decorate([ __param(0, Optional()),
        __param(1, Optional())
    ], ClrControlError);
ClrControlError.ɵfac = function ClrControlError_Factory(t) { return new (t || ClrControlError)(ɵngcc0.ɵɵdirectiveInject(ControlIdService, 8), ɵngcc0.ɵɵdirectiveInject(ControlClassService, 8), ɵngcc0.ɵɵdirectiveInject(ClrAriaLiveService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ClrControlError.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrControlError, selectors: [["clr-control-error"]], hostVars: 3, hostBindings: function ClrControlError_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("id", (ctx.controlIdService == null ? null : ctx.controlIdService.id) + "-error");
        ɵngcc0.ɵɵclassProp("clr-subtext", true);
    } }, features: [ɵngcc0.ɵɵProvidersFeature([ClrAriaLiveService])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrControlError_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrControlError, [{
        type: Component,
        args: [{
                providers: [ClrAriaLiveService],
                selector: 'clr-control-error',
                template: "\n    <ng-content></ng-content>\n    ",
                host: {
                    '[class.clr-subtext]': 'true',
                    '[id]': 'controlIdService?.id + "-error"'
                }
            }]
    }], function () { return [{ type: ControlIdService, decorators: [{
                type: Optional
            }] }, { type: ControlClassService, decorators: [{
                type: Optional
            }] }, { type: ClrAriaLiveService }, { type: ɵngcc0.ElementRef }]; }, null); })();
    return ClrControlError;
}());
export { ClrControlError };
