// @clr/angular/esm5/forms/common/helper.js
/**
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { Component, Optional } from '@angular/core';
import { ControlIdService } from './providers/control-id.service';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrControlHelper = /** @class */ (function () {
    function ClrControlHelper(controlIdService) {
        this.controlIdService = controlIdService;
    }
    ClrControlHelper.ctorParameters = function () { return [
        { type: ControlIdService, decorators: [{ type: Optional }] }
    ]; };
    ClrControlHelper = __decorate([ __param(0, Optional())
    ], ClrControlHelper);
ClrControlHelper.ɵfac = function ClrControlHelper_Factory(t) { return new (t || ClrControlHelper)(ɵngcc0.ɵɵdirectiveInject(ControlIdService, 8)); };
ClrControlHelper.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrControlHelper, selectors: [["clr-control-helper"]], hostVars: 3, hostBindings: function ClrControlHelper_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("id", (ctx.controlIdService == null ? null : ctx.controlIdService.id) + "-helper");
        ɵngcc0.ɵɵclassProp("clr-subtext", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrControlHelper_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrControlHelper, [{
        type: Component,
        args: [{
                selector: 'clr-control-helper',
                template: "\n    <ng-content></ng-content>\n    ",
                host: {
                    '[class.clr-subtext]': 'true',
                    '[id]': 'controlIdService?.id + "-helper"'
                }
            }]
    }], function () { return [{ type: ControlIdService, decorators: [{
                type: Optional
            }] }]; }, null); })();
    return ClrControlHelper;
}());
export { ClrControlHelper };
