// @clr/angular/esm5/forms/common/label.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, ElementRef, HostBinding, Input, OnDestroy, OnInit, Optional, Renderer2 } from '@angular/core';
import { ControlIdService } from './providers/control-id.service';
import { LayoutService } from './providers/layout.service';
import { NgControlService } from './providers/ng-control.service';
import * as ɵngcc0 from '@angular/core';
var ClrLabel = /** @class */ (function () {
    function ClrLabel(controlIdService, layoutService, ngControlService, renderer, el) {
        this.controlIdService = controlIdService;
        this.layoutService = layoutService;
        this.ngControlService = ngControlService;
        this.renderer = renderer;
        this.el = el;
        this.subscriptions = [];
        this.enableGrid = true;
    }
    Object.defineProperty(ClrLabel.prototype, "labelText", {
        get: function () {
            return this.el.nativeElement && this.el.nativeElement.textContent;
        },
        enumerable: true,
        configurable: true
    });
    ClrLabel.prototype.ngOnInit = function () {
        var _this = this;
        // Only add the clr-control-label if it is inside a control container
        if (this.controlIdService || this.ngControlService) {
            this.renderer.addClass(this.el.nativeElement, 'clr-control-label');
        }
        // Only set the grid column classes if we are in the right context and if they aren't already set
        if (this.enableGrid &&
            this.layoutService &&
            !this.layoutService.isVertical() &&
            this.el.nativeElement &&
            this.el.nativeElement.className.indexOf('clr-col') < 0) {
            this.renderer.addClass(this.el.nativeElement, 'clr-col-12');
            this.renderer.addClass(this.el.nativeElement, "clr-col-md-" + this.layoutService.labelSize);
        }
        if (this.controlIdService && !this.forAttr) {
            this.subscriptions.push(this.controlIdService.idChange.subscribe(function (id) { return (_this.forAttr = id); }));
        }
    };
    ClrLabel.prototype.disableGrid = function () {
        this.enableGrid = false;
    };
    ClrLabel.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
    };
    ClrLabel.ctorParameters = function () { return [
        { type: ControlIdService, decorators: [{ type: Optional }] },
        { type: LayoutService, decorators: [{ type: Optional }] },
        { type: NgControlService, decorators: [{ type: Optional }] },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    __decorate([
        HostBinding('attr.for'),
        Input('for')
    ], ClrLabel.prototype, "forAttr", void 0);
    ClrLabel = __decorate([ __param(0, Optional()),
        __param(1, Optional()),
        __param(2, Optional())
    ], ClrLabel);
ClrLabel.ɵfac = function ClrLabel_Factory(t) { return new (t || ClrLabel)(ɵngcc0.ɵɵdirectiveInject(ControlIdService, 8), ɵngcc0.ɵɵdirectiveInject(LayoutService, 8), ɵngcc0.ɵɵdirectiveInject(NgControlService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ClrLabel.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrLabel, selectors: [["label"]], hostVars: 1, hostBindings: function ClrLabel_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("for", ctx.forAttr);
    } }, inputs: { forAttr: ["for", "forAttr"] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrLabel, [{
        type: Directive,
        args: [{ selector: 'label' }]
    }], function () { return [{ type: ControlIdService, decorators: [{
                type: Optional
            }] }, { type: LayoutService, decorators: [{
                type: Optional
            }] }, { type: NgControlService, decorators: [{
                type: Optional
            }] }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { forAttr: [{
            type: HostBinding,
            args: ['attr.for']
        }, {
            type: Input,
            args: ['for']
        }] }); })();
    return ClrLabel;
}());
export { ClrLabel };
