// @clr/angular/esm2015/layout/tabs/tab-overflow-content.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
let ClrTabOverflowContent = class ClrTabOverflowContent {
};
ClrTabOverflowContent.ɵfac = function ClrTabOverflowContent_Factory(t) { return new (t || ClrTabOverflowContent)(); };
ClrTabOverflowContent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTabOverflowContent, selectors: [["clr-tab-overflow-content"]], hostVars: 2, hostBindings: function ClrTabOverflowContent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("dropdown-menu", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrTabOverflowContent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTabOverflowContent, [{
        type: Component,
        args: [{
                selector: 'clr-tab-overflow-content',
                template: `
        <ng-content></ng-content>
    `,
                host: {
                    '[class.dropdown-menu]': 'true'
                }
            }]
    }], null, null); })();
export { ClrTabOverflowContent };
