// @clr/angular/esm2015/layout/tabs/tab-content.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, Inject, Input, ViewChild } from '@angular/core';
import { IF_ACTIVE_ID, IfActiveService } from '../../utils/conditional/if-active.service';
import { AriaService } from './providers/aria.service';
import { TabsService } from './providers/tabs.service';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["tabContentProjectedRef"];
function ClrTabContent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "section", 1);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("active", ctx_r1.active);
    ɵngcc0.ɵɵproperty("id", ctx_r1.tabContentId)("hidden", !ctx_r1.active);
    ɵngcc0.ɵɵattribute("aria-labelledby", ctx_r1.ariaLabelledBy)("aria-expanded", ctx_r1.active)("aria-hidden", !ctx_r1.active);
} }
const _c1 = ["*"];
let nbTabContentComponents = 0;
let ClrTabContent = class ClrTabContent {
    constructor(ifActiveService, id, ariaService, tabsService) {
        this.ifActiveService = ifActiveService;
        this.id = id;
        this.ariaService = ariaService;
        this.tabsService = tabsService;
        if (!this.tabContentId) {
            this.tabContentId = 'clr-tab-content-' + nbTabContentComponents++;
        }
    }
    // The template must be applied on the top-down phase of view-child initialization to prevent
    // components in the content from initializing before a content container exists.
    // Some child components need their container for sizing calculations.
    /* tslint:disable:no-unused-variable */
    set templateRef(value) {
        this.viewRef = this.tabsService.tabContentViewContainer.createEmbeddedView(value);
    }
    /* tslint:enable:no-unused-variable */
    get ariaLabelledBy() {
        return this.ariaService.ariaLabelledBy;
    }
    get tabContentId() {
        return this.ariaService.ariaControls;
    }
    set tabContentId(id) {
        this.ariaService.ariaControls = id;
    }
    get active() {
        return this.ifActiveService.current === this.id;
    }
    ngOnDestroy() {
        const index = this.tabsService.tabContentViewContainer.indexOf(this.viewRef);
        if (index > -1) {
            this.tabsService.tabContentViewContainer.remove(index);
        }
    }
};
ClrTabContent.ɵfac = function ClrTabContent_Factory(t) { return new (t || ClrTabContent)(ɵngcc0.ɵɵdirectiveInject(IfActiveService), ɵngcc0.ɵɵdirectiveInject(IF_ACTIVE_ID), ɵngcc0.ɵɵdirectiveInject(AriaService), ɵngcc0.ɵɵdirectiveInject(TabsService)); };
ClrTabContent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrTabContent, selectors: [["clr-tab-content"]], viewQuery: function ClrTabContent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templateRef = _t.first);
    } }, inputs: { tabContentId: ["id", "tabContentId"] }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["tabContentProjectedRef", ""], ["role", "tabpanel", 1, "tab-content", 3, "id", "hidden"]], template: function ClrTabContent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ClrTabContent_ng_template_0_Template, 2, 7, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ClrTabContent.ctorParameters = () => [
    { type: IfActiveService },
    { type: Number, decorators: [{ type: Inject, args: [IF_ACTIVE_ID,] }] },
    { type: AriaService },
    { type: TabsService }
];
__decorate([
    ViewChild('tabContentProjectedRef', { static: true })
], ClrTabContent.prototype, "templateRef", null);
__decorate([
    Input('id')
], ClrTabContent.prototype, "tabContentId", null);
ClrTabContent = __decorate([ __param(1, Inject(IF_ACTIVE_ID))
], ClrTabContent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTabContent, [{
        type: Component,
        args: [{
                selector: 'clr-tab-content',
                template: `
    <ng-template #tabContentProjectedRef>
      <section [id]="tabContentId" role="tabpanel" class="tab-content" [class.active]="active"
               [hidden]="!active"
               [attr.aria-labelledby]="ariaLabelledBy"
               [attr.aria-expanded]="active"
               [attr.aria-hidden]="!active">
        <ng-content></ng-content>
      </section>
    </ng-template>
    `
            }]
    }], function () { return [{ type: IfActiveService }, { type: Number, decorators: [{
                type: Inject,
                args: [IF_ACTIVE_ID]
            }] }, { type: AriaService }, { type: TabsService }]; }, { tabContentId: [{
            type: Input,
            args: ['id']
        }], templateRef: [{
            type: ViewChild,
            args: ['tabContentProjectedRef', { static: true }]
        }] }); })();
export { ClrTabContent };
