// @clr/angular/esm5/utils/popover/popover-content.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { Directive, EmbeddedViewRef, Inject, Input, OnDestroy, Renderer2, TemplateRef, ViewContainerRef, AfterContentChecked, EventEmitter, } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { ClrPopoverToggleService } from './providers/popover-toggle.service';
import { ClrPopoverEventsService } from './providers/popover-events.service';
import { ClrPopoverPositionService } from './providers/popover-position.service';
import { debounceTime } from 'rxjs/operators';
// https://github.com/angular/angular/issues/20351#issuecomment-344009887
/** @dynamic */
import * as ɵngcc0 from '@angular/core';
var ClrPopoverContent = /** @class */ (function () {
    function ClrPopoverContent(document, container, template, renderer, smartPositionService, smartEventsService, smartOpenService) {
        this.document = document;
        this.container = container;
        this.template = template;
        this.renderer = renderer;
        this.smartPositionService = smartPositionService;
        this.smartEventsService = smartEventsService;
        this.smartOpenService = smartOpenService;
        this.subscriptions = [];
        this.shouldRealign = false;
        // Check-collector pattern:
        // In order to get accurate content height/width values, we cannot calculate alignment offsets until
        // after the projected content has stabilized.
        // As multiple check events may happen in the same rendering cycle, we need to collect all events
        // and only act after the content is really stable. Or we may get wrong intermediate positioning values.
        // We will channel subsequent content check events through this observable.
        this.checkCollector = new EventEmitter();
    }
    Object.defineProperty(ClrPopoverContent.prototype, "open", {
        set: function (value) {
            this.smartOpenService.open = !!value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrPopoverContent.prototype, "contentAt", {
        set: function (position) {
            this.smartPositionService.position = position;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrPopoverContent.prototype, "outsideClickClose", {
        set: function (clickToClose) {
            this.smartEventsService.outsideClickClose = !!clickToClose;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrPopoverContent.prototype, "scrollToClose", {
        set: function (scrollToClose) {
            this.smartEventsService.scrollToClose = !!scrollToClose;
        },
        enumerable: true,
        configurable: true
    });
    ClrPopoverContent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.subscriptions.push(this.smartOpenService.openChange.subscribe(function (change) {
            if (change) {
                _this.addContent();
            }
            else {
                _this.removeContent();
            }
        }), this.smartPositionService.shouldRealign.subscribe(function () {
            _this.shouldRealign = true;
            // Avoid flickering on initialization, caused by the asynchronous nature of the
            // check-collector pattern.
            if (_this.view) {
                _this.renderer.setStyle(_this.view.rootNodes[0], 'opacity', '0');
            }
        }), 
        // Here we collect subsequent synchronously received content-check events and only take action
        // at the end of the cycle. See below for details on the check-collector pattern.
        this.checkCollector.pipe(debounceTime(0)).subscribe(function () {
            _this.alignContent();
            _this.shouldRealign = false;
            if (_this.view) {
                _this.renderer.setStyle(_this.view.rootNodes[0], 'opacity', '1');
            }
        }));
    };
    ClrPopoverContent.prototype.ngOnDestroy = function () {
        this.removeContent();
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
    };
    ClrPopoverContent.prototype.removeContent = function () {
        var _this = this;
        if (!this.view) {
            return;
        }
        this.view.rootNodes.forEach(function (node) { return _this.renderer.removeChild(_this.document.body, node); });
        this.container.clear();
        delete this.view;
    };
    /**
     * TODO(matt): investigate why DebugElement retains a reference to the nodes and causes a memory leak.
     * A note about the use of appendChild/removeChild
     * The DebugElement is keeping a reference to the detached node and its unclear why.
     * This does warrant further investigation. But, since it doesn't happen in production mode
     * it is a low priority issue for now.
     */
    ClrPopoverContent.prototype.addContent = function () {
        var _this = this;
        // Create the view container
        this.view = this.container.createEmbeddedView(this.template);
        this.smartEventsService.contentRef = this.view.rootNodes[0]; // So we know where/what to set close focus on
        this.renderer.addClass(this.view.rootNodes[0], 'clr-popover-content');
        // Reset to the begining of the document to be available for sizing/positioning calculations.
        // If we add new content to the bottom it triggers changes in the layout that may lead to false anchor
        // coordinates values.
        this.renderer.setStyle(this.view.rootNodes[0], 'top', '0px');
        this.renderer.setStyle(this.view.rootNodes[0], 'left', '0px');
        // We need to hide it during the calculation phase, while it's not yet finally positioned.
        this.renderer.setStyle(this.view.rootNodes[0], 'opacity', '0');
        this.renderer.listen(this.view.rootNodes[0], 'click', function (event) {
            _this.smartOpenService.openEvent = event;
        });
        this.view.rootNodes.forEach(function (node) {
            _this.renderer.appendChild(_this.document.body, node);
        });
        // Mark for realingment on the next content-check cycle.
        this.shouldRealign = true;
    };
    ClrPopoverContent.prototype.ngAfterContentChecked = function () {
        if (this.smartOpenService.open && this.view && this.shouldRealign) {
            // Channel content-check event through the check-collector
            this.checkCollector.emit();
        }
    };
    ClrPopoverContent.prototype.alignContent = function () {
        if (!this.view) {
            return;
        }
        var positionCoords = this.smartPositionService.alignContent(this.view.rootNodes[0]);
        this.renderer.setStyle(this.view.rootNodes[0], 'top', positionCoords.yOffset + "px");
        this.renderer.setStyle(this.view.rootNodes[0], 'left', positionCoords.xOffset + "px");
    };
    ClrPopoverContent.ctorParameters = function () { return [
        { type: HTMLDocument, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: ViewContainerRef },
        { type: TemplateRef },
        { type: Renderer2 },
        { type: ClrPopoverPositionService },
        { type: ClrPopoverEventsService },
        { type: ClrPopoverToggleService }
    ]; };
    __decorate([
        Input('clrPopoverContent')
    ], ClrPopoverContent.prototype, "open", null);
    __decorate([
        Input('clrPopoverContentAt')
    ], ClrPopoverContent.prototype, "contentAt", null);
    __decorate([
        Input('clrPopoverContentOutsideClickToClose')
    ], ClrPopoverContent.prototype, "outsideClickClose", null);
    __decorate([
        Input('clrPopoverContentScrollToClose')
    ], ClrPopoverContent.prototype, "scrollToClose", null);
    ClrPopoverContent = __decorate([ __param(0, Inject(DOCUMENT))
    ], ClrPopoverContent);
ClrPopoverContent.ɵfac = function ClrPopoverContent_Factory(t) { return new (t || ClrPopoverContent)(ɵngcc0.ɵɵdirectiveInject(DOCUMENT), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ClrPopoverPositionService), ɵngcc0.ɵɵdirectiveInject(ClrPopoverEventsService), ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService)); };
ClrPopoverContent.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrPopoverContent, selectors: [["", "clrPopoverContent", ""]], inputs: { open: ["clrPopoverContent", "open"], contentAt: ["clrPopoverContentAt", "contentAt"], outsideClickClose: ["clrPopoverContentOutsideClickToClose", "outsideClickClose"], scrollToClose: ["clrPopoverContentScrollToClose", "scrollToClose"] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrPopoverContent, [{
        type: Directive,
        args: [{ selector: '[clrPopoverContent]' }]
    }], function () { return [{ type: HTMLDocument, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc0.TemplateRef }, { type: ɵngcc0.Renderer2 }, { type: ClrPopoverPositionService }, { type: ClrPopoverEventsService }, { type: ClrPopoverToggleService }]; }, { open: [{
            type: Input,
            args: ['clrPopoverContent']
        }], contentAt: [{
            type: Input,
            args: ['clrPopoverContentAt']
        }], outsideClickClose: [{
            type: Input,
            args: ['clrPopoverContentOutsideClickToClose']
        }], scrollToClose: [{
            type: Input,
            args: ['clrPopoverContentScrollToClose']
        }] }); })();
    return ClrPopoverContent;
}());
export { ClrPopoverContent };
