// @clr/angular/esm5/utils/popover/popover.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { ClrPopoverContent } from './popover-content';
import { ClrPopoverOpenCloseButton } from './popover-open-close-button';
import { ClrPopoverCloseButton } from './popover-close-button';
import { ClrPopoverAnchor } from './popover-anchor';
import * as ɵngcc0 from '@angular/core';
var ClrPopoverModuleNext = /** @class */ (function () {
    function ClrPopoverModuleNext() {
    }
ClrPopoverModuleNext.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrPopoverModuleNext });
ClrPopoverModuleNext.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrPopoverModuleNext_Factory(t) { return new (t || ClrPopoverModuleNext)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrPopoverModuleNext, { declarations: function () { return [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]; }, exports: function () { return [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrPopoverModuleNext, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent],
                exports: [ClrPopoverAnchor, ClrPopoverCloseButton, ClrPopoverOpenCloseButton, ClrPopoverContent]
            }]
    }], function () { return []; }, null); })();
    return ClrPopoverModuleNext;
}());
export { ClrPopoverModuleNext };
