// @clr/angular/esm5/layout/nav/navigation.module.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrDropdownModule } from '../../popover/dropdown/dropdown.module';
import { MainContainerWillyWonka } from './chocolate/main-container-willy-wonka';
import { NavDetectionOompaLoompa } from './chocolate/nav-detection-oompa-loompa';
import { ClrHeader } from './header';
import { ClrNavLevel } from './nav-level';
import * as ɵngcc0 from '@angular/core';
export var CLR_NAVIGATION_DIRECTIVES = [
    ClrHeader,
    ClrNavLevel,
    NavDetectionOompaLoompa,
    MainContainerWillyWonka,
];
var ClrNavigationModule = /** @class */ (function () {
    function ClrNavigationModule() {
    }
ClrNavigationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrNavigationModule });
ClrNavigationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrNavigationModule_Factory(t) { return new (t || ClrNavigationModule)(); }, imports: [[CommonModule, ClrIconModule, ClrDropdownModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrNavigationModule, { declarations: function () { return [ClrHeader,
        ClrNavLevel,
        NavDetectionOompaLoompa,
        MainContainerWillyWonka]; }, imports: function () { return [CommonModule, ClrIconModule, ClrDropdownModule]; }, exports: function () { return [ClrHeader,
        ClrNavLevel,
        NavDetectionOompaLoompa,
        MainContainerWillyWonka]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrNavigationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule, ClrDropdownModule],
                declarations: [CLR_NAVIGATION_DIRECTIVES],
                exports: [CLR_NAVIGATION_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrNavigationModule;
}());
export { ClrNavigationModule };
