// @clr/angular/esm5/utils/conditional/if-open.directive.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, EventEmitter, Input, OnDestroy, Output, TemplateRef, ViewContainerRef } from '@angular/core';
import { ClrPopoverToggleService } from '../popover/providers/popover-toggle.service';
import * as ɵngcc0 from '@angular/core';
var ClrIfOpen = /** @class */ (function () {
    function ClrIfOpen(toggleService, template, container) {
        var _this = this;
        this.toggleService = toggleService;
        this.template = template;
        this.container = container;
        /**********
         * @property openChange
         *
         * @description
         * An event emitter that emits when the open property is set to allow for 2way binding when the directive is
         * used with de-structured / de-sugared syntax.
         */
        this.openChange = new EventEmitter(false);
        this.subscription = this.toggleService.openChange.subscribe(function (change) {
            _this.updateView(change);
            _this.openChange.emit(change);
        });
    }
    Object.defineProperty(ClrIfOpen.prototype, "open", {
        /********
         *
         * @description
         * A getter that returns the current ClrPopoverToggleService.open value.
         *
         */
        get: function () {
            return this.toggleService.open;
        },
        /*********
         *
         * @description
         * A setter that updates ClrPopoverToggleService.open with value.
         *
         * @param value
         */
        set: function (value) {
            this.toggleService.open = value;
        },
        enumerable: true,
        configurable: true
    });
    /*********
     *
     * @description
     * Function that takes a boolean value and either created an embedded view for the associated ViewContainerRef or,
     * Clears all views from the ViewContainerRef
     * @param value
     */
    ClrIfOpen.prototype.updateView = function (value) {
        if (value) {
            this.container.createEmbeddedView(this.template);
        }
        else {
            this.container.clear();
        }
    };
    ClrIfOpen.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    ClrIfOpen.ctorParameters = function () { return [
        { type: ClrPopoverToggleService },
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    __decorate([
        Input('clrIfOpen')
    ], ClrIfOpen.prototype, "open", null);
    __decorate([
        Output('clrIfOpenChange')
    ], ClrIfOpen.prototype, "openChange", void 0);
ClrIfOpen.ɵfac = function ClrIfOpen_Factory(t) { return new (t || ClrIfOpen)(ɵngcc0.ɵɵdirectiveInject(ClrPopoverToggleService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
ClrIfOpen.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClrIfOpen, selectors: [["", "clrIfOpen", ""]], inputs: { open: ["clrIfOpen", "open"] }, outputs: { openChange: "clrIfOpenChange" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrIfOpen, [{
        type: Directive,
        args: [{ selector: '[clrIfOpen]' }]
    }], function () { return [{ type: ClrPopoverToggleService }, { type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }]; }, { openChange: [{
            type: Output,
            args: ['clrIfOpenChange']
        }], open: [{
            type: Input,
            args: ['clrIfOpen']
        }] }); })();
    return ClrIfOpen;
}());
export { ClrIfOpen };
