// @clr/angular/esm5/utils/focus/key-focus/key-focus.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrKeyFocus } from './key-focus';
import { ClrKeyFocusItem } from './key-focus-item';
import * as ɵngcc0 from '@angular/core';
var KEY_FOCUS_DIRECTIVES = [ClrKeyFocus, ClrKeyFocusItem];
var ClrKeyFocusModule = /** @class */ (function () {
    function ClrKeyFocusModule() {
    }
ClrKeyFocusModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrKeyFocusModule });
ClrKeyFocusModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrKeyFocusModule_Factory(t) { return new (t || ClrKeyFocusModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrKeyFocusModule, { declarations: function () { return [ClrKeyFocus, ClrKeyFocusItem]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrKeyFocus, ClrKeyFocusItem]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrKeyFocusModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [KEY_FOCUS_DIRECTIVES],
                exports: [KEY_FOCUS_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrKeyFocusModule;
}());
export { ClrKeyFocusModule };
