// @clr/angular/esm5/forms/password/password.module.js
/**
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrPassword } from './password';
import { ClrPasswordContainer } from './password-container';
import * as ɵngcc0 from '@angular/core';
var ClrPasswordModule = /** @class */ (function () {
    function ClrPasswordModule() {
    }
ClrPasswordModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrPasswordModule });
ClrPasswordModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrPasswordModule_Factory(t) { return new (t || ClrPasswordModule)(); }, imports: [[CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrPasswordModule, { declarations: function () { return [ClrPassword, ClrPasswordContainer]; }, imports: function () { return [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule]; }, exports: function () { return [ClrCommonFormsModule, ClrPassword, ClrPasswordContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrPasswordModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
                declarations: [ClrPassword, ClrPasswordContainer],
                exports: [ClrCommonFormsModule, ClrPassword, ClrPasswordContainer],
                entryComponents: [ClrPasswordContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrPasswordModule;
}());
export { ClrPasswordModule };
