// @clr/angular/esm5/forms/forms.module.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrCheckboxModule } from './checkbox/checkbox.module';
import { ClrCommonFormsModule } from './common/common.module';
import { ClrDatepickerModule } from './datepicker/datepicker.module';
import { ClrInputModule } from './input/input.module';
import { ClrPasswordModule } from './password/password.module';
import { ClrRadioModule } from './radio/radio.module';
import { ClrSelectModule } from './select/select.module';
import { ClrTextareaModule } from './textarea/textarea.module';
import { ClrRangeModule } from './range/range.module';
import { ClrDatalistModule } from './datalist/datalist.module';
import * as ɵngcc0 from '@angular/core';
var ClrFormsModule = /** @class */ (function () {
    function ClrFormsModule() {
    }
ClrFormsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrFormsModule });
ClrFormsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrFormsModule_Factory(t) { return new (t || ClrFormsModule)(); }, imports: [[CommonModule],
        ClrCommonFormsModule,
        ClrCheckboxModule,
        ClrDatepickerModule,
        ClrInputModule,
        ClrPasswordModule,
        ClrRadioModule,
        ClrSelectModule,
        ClrTextareaModule,
        ClrRangeModule,
        ClrDatalistModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrFormsModule, { imports: function () { return [CommonModule]; }, exports: function () { return [ClrCommonFormsModule,
        ClrCheckboxModule,
        ClrDatepickerModule,
        ClrInputModule,
        ClrPasswordModule,
        ClrRadioModule,
        ClrSelectModule,
        ClrTextareaModule,
        ClrRangeModule,
        ClrDatalistModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrFormsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [
                    ClrCommonFormsModule,
                    ClrCheckboxModule,
                    ClrDatepickerModule,
                    ClrInputModule,
                    ClrPasswordModule,
                    ClrRadioModule,
                    ClrSelectModule,
                    ClrTextareaModule,
                    ClrRangeModule,
                    ClrDatalistModule,
                ]
            }]
    }], function () { return []; }, null); })();
    return ClrFormsModule;
}());
export { ClrFormsModule };
