// @clr/angular/esm2015/accordion/stepper/chocolate/stepper-oompa-loompa.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __param } from "tslib";
import { ChangeDetectorRef, Directive, Optional } from '@angular/core';
import { OompaLoompa } from '../../../utils/chocolate/oompa-loompa';
import { StepperWillyWonka } from './stepper-willy-wonka';
import { IfExpandService } from '../../../utils/conditional/if-expanded.service';
import * as ɵngcc0 from '@angular/core';
let StepperOompaLoompa = class StepperOompaLoompa extends OompaLoompa {
    constructor(cdr, willyWonka, ifExpandService) {
        if (!willyWonka) {
            throw new Error('clr-stepper-panel should only be used inside of clrStepper');
        }
        super(cdr, willyWonka);
        this.expand = ifExpandService;
    }
    get flavor() {
        return this.expand.expanded;
    }
};
StepperOompaLoompa.ɵfac = function StepperOompaLoompa_Factory(t) { return new (t || StepperOompaLoompa)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(StepperWillyWonka, 8), ɵngcc0.ɵɵdirectiveInject(IfExpandService)); };
StepperOompaLoompa.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: StepperOompaLoompa, selectors: [["clr-stepper-panel"], ["", "clrStepButton", ""]], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
StepperOompaLoompa.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: StepperWillyWonka, decorators: [{ type: Optional }] },
    { type: IfExpandService }
];
StepperOompaLoompa = __decorate([ __param(1, Optional())
], StepperOompaLoompa);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StepperOompaLoompa, [{
        type: Directive,
        args: [{ selector: 'clr-stepper-panel, [clrStepButton]' }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: StepperWillyWonka, decorators: [{
                type: Optional
            }] }, { type: IfExpandService }]; }, null); })();
export { StepperOompaLoompa };
