// @clr/angular/esm5/utils/animations/expandable-animation/expandable-animation.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, ElementRef, HostBinding, HostListener, Input } from '@angular/core';
import { animate, style, transition, trigger } from '@angular/animations';
import { DomAdapter } from '../../dom-adapter/dom-adapter';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrExpandableAnimation = /** @class */ (function () {
    function ClrExpandableAnimation(element, domAdapter) {
        this.element = element;
        this.domAdapter = domAdapter;
        this.startHeight = 0;
    }
    Object.defineProperty(ClrExpandableAnimation.prototype, "expandAnimation", {
        get: function () {
            return { value: this.clrExpandTrigger, params: { startHeight: this.startHeight } };
        },
        enumerable: true,
        configurable: true
    });
    ClrExpandableAnimation.prototype.animationDone = function () {
        // A "safe" auto-update of the height ensuring basic OOTB user experience .
        // Prone to small jumps in initial animation height if data was changed in the meantime, window was resized, etc.
        // For optimal behavior call manually updateStartHeight() from the parent component before initiating the update.
        this.updateStartHeight();
    };
    ClrExpandableAnimation.prototype.updateStartHeight = function () {
        this.startHeight = this.domAdapter.computedHeight(this.element.nativeElement) || 0;
    };
    ClrExpandableAnimation.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DomAdapter }
    ]; };
    __decorate([
        Input()
    ], ClrExpandableAnimation.prototype, "clrExpandTrigger", void 0);
    __decorate([
        HostBinding('@expandAnimation')
    ], ClrExpandableAnimation.prototype, "expandAnimation", null);
    __decorate([
        HostListener('@expandAnimation.done')
    ], ClrExpandableAnimation.prototype, "animationDone", null);
ClrExpandableAnimation.ɵfac = function ClrExpandableAnimation_Factory(t) { return new (t || ClrExpandableAnimation)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(DomAdapter)); };
ClrExpandableAnimation.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrExpandableAnimation, selectors: [["clr-expandable-animation"]], hostVars: 1, hostBindings: function ClrExpandableAnimation_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵcomponentHostSyntheticListener("@expandAnimation.done", function ClrExpandableAnimation_animation_expandAnimation_done_HostBindingHandler() { return ctx.animationDone(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵupdateSyntheticHostBinding("@expandAnimation", ctx.expandAnimation);
    } }, inputs: { clrExpandTrigger: "clrExpandTrigger" }, features: [ɵngcc0.ɵɵProvidersFeature([DomAdapter])], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrExpandableAnimation_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, styles: ["[_nghost-%COMP%] {\n      display: block;\n    }"], data: { animation: [
            trigger('expandAnimation', [
                transition('true <=> false', [
                    style({ height: '{{startHeight}}px', overflow: 'hidden' }),
                    animate('0.2s ease-in-out', style({ height: '*' })),
                ]),
            ]),
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrExpandableAnimation, [{
        type: Component,
        args: [{
                selector: 'clr-expandable-animation',
                template: "\n    <ng-content></ng-content>\n  ",
                animations: [
                    trigger('expandAnimation', [
                        transition('true <=> false', [
                            style({ height: '{{startHeight}}px', overflow: 'hidden' }),
                            animate('0.2s ease-in-out', style({ height: '*' })),
                        ]),
                    ]),
                ],
                providers: [DomAdapter],
                styles: ["\n    :host {\n      display: block;\n    }\n  "]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: DomAdapter }]; }, { expandAnimation: [{
            type: HostBinding,
            args: ['@expandAnimation']
        }], animationDone: [{
            type: HostListener,
            args: ['@expandAnimation.done']
        }], clrExpandTrigger: [{
            type: Input
        }] }); })();
    return ClrExpandableAnimation;
}());
export { ClrExpandableAnimation };
