// @clr/angular/esm5/utils/animations/expandable-animation/expandable-animation.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EXPANDABLE_ANIMATION_DIRECTIVES } from './index';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './expandable-animation';
var ClrExpandableAnimationModule = /** @class */ (function () {
    function ClrExpandableAnimationModule() {
    }
ClrExpandableAnimationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrExpandableAnimationModule });
ClrExpandableAnimationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrExpandableAnimationModule_Factory(t) { return new (t || ClrExpandableAnimationModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrExpandableAnimationModule, { declarations: function () { return [ɵngcc1.ClrExpandableAnimation]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ɵngcc1.ClrExpandableAnimation]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrExpandableAnimationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [EXPANDABLE_ANIMATION_DIRECTIVES],
                exports: [EXPANDABLE_ANIMATION_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrExpandableAnimationModule;
}());
export { ClrExpandableAnimationModule };
