// @clr/angular/esm5/forms/textarea/textarea.module.js
/**
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrCommonFormsModule } from '../common/common.module';
import { ClrTextarea } from './textarea';
import { ClrTextareaContainer } from './textarea-container';
import * as ɵngcc0 from '@angular/core';
var ClrTextareaModule = /** @class */ (function () {
    function ClrTextareaModule() {
    }
ClrTextareaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTextareaModule });
ClrTextareaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrTextareaModule_Factory(t) { return new (t || ClrTextareaModule)(); }, imports: [[CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
        ClrCommonFormsModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTextareaModule, { declarations: function () { return [ClrTextarea, ClrTextareaContainer]; }, imports: function () { return [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule]; }, exports: function () { return [ClrCommonFormsModule, ClrTextarea, ClrTextareaContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTextareaModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, ClrIconModule, ClrCommonFormsModule],
                declarations: [ClrTextarea, ClrTextareaContainer],
                exports: [ClrCommonFormsModule, ClrTextarea, ClrTextareaContainer],
                entryComponents: [ClrTextareaContainer]
            }]
    }], function () { return []; }, null); })();
    return ClrTextareaModule;
}());
export { ClrTextareaModule };
