// @clr/angular/esm2015/forms/datepicker/providers/date-form-control.service.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
let DateFormControlService = class DateFormControlService {
    constructor() {
        this._touchedChange = new Subject();
        this._dirtyChange = new Subject();
    }
    get touchedChange() {
        return this._touchedChange.asObservable();
    }
    get dirtyChange() {
        return this._dirtyChange.asObservable();
    }
    markAsTouched() {
        this._touchedChange.next();
    }
    markAsDirty() {
        this._dirtyChange.next();
    }
    // friendly wrapper
    setDisabled(state) {
        this.disabled = state;
    }
};
DateFormControlService.ɵfac = function DateFormControlService_Factory(t) { return new (t || DateFormControlService)(); };
DateFormControlService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: DateFormControlService, factory: DateFormControlService.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DateFormControlService, [{
        type: Injectable
    }], function () { return []; }, null); })();
export { DateFormControlService };
