// @clr/angular/esm2015/forms/datepicker/providers/view-manager.service.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ClrPopoverPositions } from '../../../utils/popover/enums/positions.enum';
/**
 * This service manages which view is visible in the datepicker popover.
 */
import * as ɵngcc0 from '@angular/core';
let ViewManagerService = class ViewManagerService {
    constructor() {
        this._currentView = "DAYVIEW" /* DAYVIEW */;
        this.position = ClrPopoverPositions['bottom-left'];
    }
    get isDayView() {
        return this._currentView === "DAYVIEW" /* DAYVIEW */;
    }
    get isYearView() {
        return this._currentView === "YEARVIEW" /* YEARVIEW */;
    }
    get isMonthView() {
        return this._currentView === "MONTHVIEW" /* MONTHVIEW */;
    }
    changeToMonthView() {
        this._currentView = "MONTHVIEW" /* MONTHVIEW */;
    }
    changeToYearView() {
        this._currentView = "YEARVIEW" /* YEARVIEW */;
    }
    changeToDayView() {
        this._currentView = "DAYVIEW" /* DAYVIEW */;
    }
};
ViewManagerService.ɵfac = function ViewManagerService_Factory(t) { return new (t || ViewManagerService)(); };
ViewManagerService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ViewManagerService, factory: ViewManagerService.ɵfac });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ViewManagerService, [{
        type: Injectable
    }], function () { return []; }, null); })();
export { ViewManagerService };
