// @clr/angular/esm5/popover/tooltip/tooltip.module.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrConditionalModule } from '../../utils/conditional/conditional.module';
import { ClrTooltip } from './tooltip';
import { ClrTooltipContent } from './tooltip-content';
import { ClrTooltipTrigger } from './tooltip-trigger';
import * as ɵngcc0 from '@angular/core';
export var CLR_TOOLTIP_DIRECTIVES = [ClrTooltip, ClrTooltipTrigger, ClrTooltipContent];
var ClrTooltipModule = /** @class */ (function () {
    function ClrTooltipModule() {
    }
ClrTooltipModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrTooltipModule });
ClrTooltipModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrTooltipModule_Factory(t) { return new (t || ClrTooltipModule)(); }, imports: [[CommonModule],
        ClrConditionalModule, ClrIconModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrTooltipModule, { declarations: function () { return [ClrTooltip, ClrTooltipTrigger, ClrTooltipContent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrTooltip, ClrTooltipTrigger, ClrTooltipContent,
        ClrConditionalModule, ClrIconModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrTooltipModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_TOOLTIP_DIRECTIVES],
                exports: [CLR_TOOLTIP_DIRECTIVES, ClrConditionalModule, ClrIconModule]
            }]
    }], function () { return []; }, null); })();
    return ClrTooltipModule;
}());
export { ClrTooltipModule };
