// @clr/angular/esm2015/layout/tabs/chocolate/active-oompa-loompa.js
import { __decorate, __param } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ChangeDetectorRef, Directive, Inject, Optional } from '@angular/core';
import { OompaLoompa } from '../../../utils/chocolate/oompa-loompa';
import { IF_ACTIVE_ID, IfActiveService } from '../../../utils/conditional/if-active.service';
import { TabsWillyWonka } from './tabs-willy-wonka';
import * as ɵngcc0 from '@angular/core';
let ActiveOompaLoompa = class ActiveOompaLoompa extends OompaLoompa {
    constructor(cdr, willyWonka, id, ifActive) {
        if (!willyWonka) {
            throw new Error('clrTabLink and clr-tab-content should only be used inside of a clr-tabs');
        }
        super(cdr, willyWonka);
        this.ifActive = ifActive;
        this.id = id;
    }
    get flavor() {
        return this.ifActive.current === this.id;
    }
};
ActiveOompaLoompa.ɵfac = function ActiveOompaLoompa_Factory(t) { return new (t || ActiveOompaLoompa)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(TabsWillyWonka, 8), ɵngcc0.ɵɵdirectiveInject(IF_ACTIVE_ID), ɵngcc0.ɵɵdirectiveInject(IfActiveService)); };
ActiveOompaLoompa.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ActiveOompaLoompa, selectors: [["", "clrTabLink", ""], ["clr-tab-content"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
ActiveOompaLoompa.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: TabsWillyWonka, decorators: [{ type: Optional }] },
    { type: Number, decorators: [{ type: Inject, args: [IF_ACTIVE_ID,] }] },
    { type: IfActiveService }
];
ActiveOompaLoompa = __decorate([ __param(1, Optional()),
    __param(2, Inject(IF_ACTIVE_ID))
], ActiveOompaLoompa);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ActiveOompaLoompa, [{
        type: Directive,
        args: [{ selector: '[clrTabLink], clr-tab-content' }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: TabsWillyWonka, decorators: [{
                type: Optional
            }] }, { type: Number, decorators: [{
                type: Inject,
                args: [IF_ACTIVE_ID]
            }] }, { type: IfActiveService }]; }, null); })();
export { ActiveOompaLoompa };
