// @clr/angular/esm5/accordion/accordion.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, ContentChildren, ChangeDetectionStrategy, Input, } from '@angular/core';
import { startWith } from 'rxjs/operators';
import { AccordionService } from './providers/accordion.service';
import { ClrAccordionPanel } from './accordion-panel';
import { AccordionStrategy } from './enums/accordion-strategy.enum';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrAccordion = /** @class */ (function () {
    function ClrAccordion(accordionService) {
        this.accordionService = accordionService;
        this.multiPanel = false;
        this.subscriptions = [];
    }
    ClrAccordion.prototype.ngOnInit = function () {
        this.setAccordionStrategy();
    };
    ClrAccordion.prototype.ngOnChanges = function (changes) {
        if (changes.multiPanel.currentValue !== changes.multiPanel.previousValue) {
            this.setAccordionStrategy();
        }
    };
    ClrAccordion.prototype.ngAfterViewInit = function () {
        this.subscriptions.push(this.listenForDOMChanges());
    };
    ClrAccordion.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (s) { return s.unsubscribe(); });
    };
    ClrAccordion.prototype.setAccordionStrategy = function () {
        var strategy = this.multiPanel ? AccordionStrategy.Multi : AccordionStrategy.Default;
        this.accordionService.setStrategy(strategy);
    };
    ClrAccordion.prototype.listenForDOMChanges = function () {
        var _this = this;
        return this.panels.changes
            .pipe(startWith(this.panels))
            .subscribe(function (panels) { return _this.accordionService.updatePanelOrder(panels.toArray().map(function (p) { return p.id; })); });
    };
    ClrAccordion.ctorParameters = function () { return [
        { type: AccordionService }
    ]; };
    __decorate([
        Input('clrAccordionMultiPanel')
    ], ClrAccordion.prototype, "multiPanel", void 0);
    __decorate([
        ContentChildren(ClrAccordionPanel, { descendants: true })
    ], ClrAccordion.prototype, "panels", void 0);
ClrAccordion.ɵfac = function ClrAccordion_Factory(t) { return new (t || ClrAccordion)(ɵngcc0.ɵɵdirectiveInject(AccordionService)); };
ClrAccordion.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAccordion, selectors: [["clr-accordion"]], contentQueries: function ClrAccordion_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ClrAccordionPanel, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.panels = _t);
    } }, hostVars: 2, hostBindings: function ClrAccordion_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("clr-accordion", true);
    } }, inputs: { multiPanel: ["clrAccordionMultiPanel", "multiPanel"] }, features: [ɵngcc0.ɵɵProvidersFeature([AccordionService]), ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrAccordion_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAccordion, [{
        type: Component,
        args: [{
                selector: 'clr-accordion',
                template: "<ng-content></ng-content>",
                host: { '[class.clr-accordion]': 'true' },
                providers: [AccordionService],
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: AccordionService }]; }, { multiPanel: [{
            type: Input,
            args: ['clrAccordionMultiPanel']
        }], panels: [{
            type: ContentChildren,
            args: [ClrAccordionPanel, { descendants: true }]
        }] }); })();
    return ClrAccordion;
}());
export { ClrAccordion };
