// @clr/angular/esm5/accordion/accordion.module.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate, __read, __spread } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../icon/icon.module';
import { ClrAccordionPanel } from './accordion-panel';
import { ClrAccordionTitle } from './accordion-title';
import { ClrAccordionDescription } from './accordion-description';
import { ClrAccordion } from './accordion';
import { ClrAccordionContent } from './accordion-content';
import { AccordionOompaLoompa } from './chocolate/accordion-oompa-loompa';
import { AccordionWillyWonka } from './chocolate/accordion-willy-wonka';
import * as ɵngcc0 from '@angular/core';
var declarations = [
    ClrAccordion,
    ClrAccordionPanel,
    ClrAccordionTitle,
    ClrAccordionDescription,
    ClrAccordionContent,
    AccordionOompaLoompa,
    AccordionWillyWonka,
];
var ClrAccordionModule = /** @class */ (function () {
    function ClrAccordionModule() {
    }
ClrAccordionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrAccordionModule });
ClrAccordionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrAccordionModule_Factory(t) { return new (t || ClrAccordionModule)(); }, imports: [[CommonModule, ClrIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrAccordionModule, { declarations: function () { return [ClrAccordion,
        ClrAccordionPanel,
        ClrAccordionTitle,
        ClrAccordionDescription,
        ClrAccordionContent,
        AccordionOompaLoompa,
        AccordionWillyWonka]; }, imports: function () { return [CommonModule, ClrIconModule]; }, exports: function () { return [ClrAccordion,
        ClrAccordionPanel,
        ClrAccordionTitle,
        ClrAccordionDescription,
        ClrAccordionContent,
        AccordionOompaLoompa,
        AccordionWillyWonka]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAccordionModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule],
                declarations: __spread(declarations),
                exports: __spread(declarations)
            }]
    }], function () { return []; }, null); })();
    return ClrAccordionModule;
}());
export { ClrAccordionModule };
