// @clr/angular/esm5/accordion/accordion-content.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var ClrAccordionContent = /** @class */ (function () {
    function ClrAccordionContent() {
    }
ClrAccordionContent.ɵfac = function ClrAccordionContent_Factory(t) { return new (t || ClrAccordionContent)(); };
ClrAccordionContent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClrAccordionContent, selectors: [["clr-accordion-content"], ["clr-step-content"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ClrAccordionContent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrAccordionContent, [{
        type: Component,
        args: [{
                selector: 'clr-accordion-content, clr-step-content',
                template: "<ng-content></ng-content>",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, null); })();
    return ClrAccordionContent;
}());
export { ClrAccordionContent };
