// @clr/angular/esm5/progress/spinner/spinner.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrSpinner } from './spinner';
import * as ɵngcc0 from '@angular/core';
export var CLR_SPINNER_DIRECTIVES = [ClrSpinner];
var ClrSpinnerModule = /** @class */ (function () {
    function ClrSpinnerModule() {
    }
ClrSpinnerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrSpinnerModule });
ClrSpinnerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrSpinnerModule_Factory(t) { return new (t || ClrSpinnerModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrSpinnerModule, { declarations: function () { return [ClrSpinner]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ClrSpinner]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrSpinnerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [CLR_SPINNER_DIRECTIVES],
                exports: [CLR_SPINNER_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrSpinnerModule;
}());
export { ClrSpinnerModule };
