// @clr/angular/esm5/layout/main-container/main-container.module.js
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrIconModule } from '../../icon/icon.module';
import { ClrMainContainer } from './main-container';
import * as ɵngcc0 from '@angular/core';
export var CLR_LAYOUT_DIRECTIVES = [ClrMainContainer];
var ClrMainContainerModule = /** @class */ (function () {
    function ClrMainContainerModule() {
    }
ClrMainContainerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClrMainContainerModule });
ClrMainContainerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClrMainContainerModule_Factory(t) { return new (t || ClrMainContainerModule)(); }, imports: [[CommonModule, ClrIconModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClrMainContainerModule, { declarations: function () { return [ClrMainContainer]; }, imports: function () { return [CommonModule, ClrIconModule]; }, exports: function () { return [ClrMainContainer]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClrMainContainerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ClrIconModule],
                declarations: [CLR_LAYOUT_DIRECTIVES],
                exports: [CLR_LAYOUT_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return ClrMainContainerModule;
}());
export { ClrMainContainerModule };
