// @clr/angular/esm5/utils/outside-click/outside-click.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, ElementRef, EventEmitter, HostListener, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var OutsideClick = /** @class */ (function () {
    function OutsideClick(el) {
        this.el = el;
        this.strict = false;
        this.outsideClick = new EventEmitter(false);
    }
    OutsideClick.prototype.documentClick = function (event) {
        var target = event.target; // Get the element in the DOM on which the mouse was clicked
        var host = this.el.nativeElement; // Get the current actionMenu native HTML element
        if (target === host) {
            return;
        }
        if (!this.strict && host.contains(target)) {
            return;
        }
        this.outsideClick.emit(event);
    };
    OutsideClick.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Input('clrStrict')
    ], OutsideClick.prototype, "strict", void 0);
    __decorate([
        Output('clrOutsideClick')
    ], OutsideClick.prototype, "outsideClick", void 0);
    __decorate([
        HostListener('document:click', ['$event'])
    ], OutsideClick.prototype, "documentClick", null);
OutsideClick.ɵfac = function OutsideClick_Factory(t) { return new (t || OutsideClick)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
OutsideClick.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: OutsideClick, selectors: [["", "clrOutsideClick", ""]], hostBindings: function OutsideClick_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function OutsideClick_click_HostBindingHandler($event) { return ctx.documentClick($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { strict: ["clrStrict", "strict"] }, outputs: { outsideClick: "clrOutsideClick" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(OutsideClick, [{
        type: Directive,
        args: [{ selector: '[clrOutsideClick]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { strict: [{
            type: Input,
            args: ['clrStrict']
        }], outsideClick: [{
            type: Output,
            args: ['clrOutsideClick']
        }], documentClick: [{
            type: HostListener,
            args: ['document:click', ['$event']]
        }] }); })();
    return OutsideClick;
}());
export { OutsideClick };
